// @phantom/client@2.0.1/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  AddressType: () => import_openapi_wallet_service3.DerivationInfoAddressFormatEnum,
  DerivationPath: () => DerivationPath,
  NetworkId: () => import_constants4.NetworkId,
  PhantomClient: () => PhantomClient,
  SpendingLimitError: () => SpendingLimitError,
  TransactionBlockedError: () => TransactionBlockedError,
  WalletServiceError: () => WalletServiceError,
  deriveSubmissionConfig: () => deriveSubmissionConfig,
  generateKeyPair: () => import_crypto.generateKeyPair,
  getAxiosErrorData: () => getAxiosErrorData,
  getDerivationPathForNetwork: () => getDerivationPathForNetwork,
  getErrorMessage: () => getErrorMessage,
  getNetworkConfig: () => getNetworkConfig,
  getNetworkDescription: () => getNetworkDescription,
  getNetworkIdsByChain: () => getNetworkIdsByChain,
  getSupportedNetworkIds: () => getSupportedNetworkIds,
  isAuthenticationError: () => isAuthenticationError,
  parseWalletServiceError: () => parseWalletServiceError,
  supportsTransactionSubmission: () => supportsTransactionSubmission
});
module.exports = __toCommonJS(src_exports);

// src/PhantomClient.ts
var import_openapi_wallet_service2 = require("@phantom/openapi-wallet-service");
var import_axios2 = __toESM(require("axios"));
var import_buffer = require("buffer");

// src/caip2-mappings.ts
var import_constants = require("@phantom/constants");
var CAIP2_NETWORK_MAPPINGS = {
  // Solana networks
  [import_constants.NetworkId.SOLANA_MAINNET]: {
    chain: "solana",
    network: "mainnet",
    description: "Solana Mainnet-Beta"
  },
  [import_constants.NetworkId.SOLANA_DEVNET]: {
    chain: "solana",
    network: "devnet",
    description: "Solana Devnet"
  },
  [import_constants.NetworkId.SOLANA_TESTNET]: {
    chain: "solana",
    network: "testnet",
    description: "Solana Testnet"
  },
  // Ethereum/EVM networks
  [import_constants.NetworkId.ETHEREUM_MAINNET]: {
    chain: "ethereum",
    network: "mainnet",
    description: "Ethereum Mainnet"
  },
  [import_constants.NetworkId.ETHEREUM_SEPOLIA]: {
    chain: "ethereum",
    network: "sepolia",
    description: "Sepolia Testnet"
  },
  [import_constants.NetworkId.POLYGON_MAINNET]: {
    chain: "polygon",
    network: "mainnet",
    description: "Polygon Mainnet"
  },
  [import_constants.NetworkId.POLYGON_AMOY]: {
    chain: "polygon",
    network: "amoy",
    description: "Polygon Amoy Testnet"
  },
  [import_constants.NetworkId.BASE_MAINNET]: {
    chain: "base",
    network: "mainnet",
    description: "Base Mainnet"
  },
  [import_constants.NetworkId.BASE_SEPOLIA]: {
    chain: "base",
    network: "sepolia",
    description: "Base Sepolia Testnet"
  },
  [import_constants.NetworkId.ARBITRUM_ONE]: {
    chain: "arbitrum",
    network: "mainnet",
    description: "Arbitrum One"
  },
  [import_constants.NetworkId.ARBITRUM_SEPOLIA]: {
    chain: "arbitrum",
    network: "sepolia",
    description: "Arbitrum Sepolia Testnet"
  },
  [import_constants.NetworkId.MONAD_MAINNET]: {
    chain: "monad",
    network: "mainnet",
    description: "Monad Mainnet"
  },
  [import_constants.NetworkId.MONAD_TESTNET]: {
    chain: "monad",
    network: "testnet",
    description: "Monad Testnet"
  },
  // Bitcoin networks (for future support)
  [import_constants.NetworkId.BITCOIN_MAINNET]: {
    chain: "bitcoin",
    network: "mainnet",
    description: "Bitcoin Mainnet"
  },
  [import_constants.NetworkId.BITCOIN_TESTNET]: {
    chain: "bitcoin",
    network: "testnet",
    description: "Bitcoin Testnet"
  },
  // Sui networks (for future support)
  [import_constants.NetworkId.SUI_MAINNET]: {
    chain: "sui",
    network: "mainnet",
    description: "Sui Mainnet"
  },
  [import_constants.NetworkId.SUI_TESTNET]: {
    chain: "sui",
    network: "testnet",
    description: "Sui Testnet"
  }
};
function deriveSubmissionConfig(networkId) {
  const mapping = CAIP2_NETWORK_MAPPINGS[networkId];
  if (!mapping) {
    return void 0;
  }
  return {
    chain: mapping.chain,
    network: mapping.network
  };
}
function supportsTransactionSubmission(networkId) {
  return networkId in CAIP2_NETWORK_MAPPINGS;
}
function getNetworkDescription(networkId) {
  return CAIP2_NETWORK_MAPPINGS[networkId]?.description;
}
function getSupportedNetworkIds() {
  return Object.keys(CAIP2_NETWORK_MAPPINGS);
}
function getNetworkIdsByChain(chain) {
  return Object.entries(CAIP2_NETWORK_MAPPINGS).filter(([_, mapping]) => mapping.chain.toLowerCase() === chain.toLowerCase()).map(([networkId]) => networkId);
}

// src/constants.ts
var import_openapi_wallet_service = require("@phantom/openapi-wallet-service");
var DerivationPath = {
  // Solana - BIP44 standard for Solana (coin type 501)
  Solana: (accountIndex = 0) => `m/44'/501'/${accountIndex}'/0'`,
  // Ethereum - BIP44 standard for Ethereum and all EVM-compatible chains (coin type 60)
  Ethereum: (accountIndex = 0) => `m/44'/60'/0'/0/${accountIndex}`,
  // Bitcoin - BIP84 standard for Bitcoin (coin type 0)
  Bitcoin: (accountIndex = 0) => `m/84'/0'/${accountIndex}'/0`,
  // Sui - BIP44 standard for Sui (coin type 784)
  Sui: (accountIndex = 0) => `m/44'/784'/${accountIndex}'/0'/0'`
};
function getDerivationPathForNetwork(networkId, accountIndex = 0) {
  const network = networkId.split(":")[0].toLowerCase();
  switch (network) {
    case "solana":
      return DerivationPath.Solana(accountIndex);
    case "sui":
      return DerivationPath.Sui(accountIndex);
    case "bitcoin":
    case "btc":
      return DerivationPath.Bitcoin(accountIndex);
    case "eip155":
    case "ethereum":
    case "eth":
    default:
      return DerivationPath.Ethereum(accountIndex);
  }
}
function getNetworkConfig(networkId, accountIndex = 0) {
  const network = networkId.split(":")[0].toLowerCase();
  switch (network) {
    case "solana":
      return {
        derivationPath: DerivationPath.Solana(accountIndex),
        curve: import_openapi_wallet_service.DerivationInfoCurveEnum.ed25519,
        algorithm: import_openapi_wallet_service.Algorithm.ed25519,
        addressFormat: import_openapi_wallet_service.DerivationInfoAddressFormatEnum.solana
      };
    case "sui":
      return {
        derivationPath: DerivationPath.Sui(accountIndex),
        curve: import_openapi_wallet_service.DerivationInfoCurveEnum.ed25519,
        algorithm: import_openapi_wallet_service.Algorithm.ed25519,
        addressFormat: import_openapi_wallet_service.DerivationInfoAddressFormatEnum.sui
      };
    case "bitcoin":
    case "btc":
      return {
        derivationPath: DerivationPath.Bitcoin(accountIndex),
        curve: import_openapi_wallet_service.DerivationInfoCurveEnum.secp256k1,
        algorithm: import_openapi_wallet_service.Algorithm.secp256k1,
        addressFormat: import_openapi_wallet_service.DerivationInfoAddressFormatEnum.bitcoinSegwit
        // Bitcoin uses a different format, but for SDK consistency we use Ethereum format
      };
    case "eip155":
      return {
        derivationPath: DerivationPath.Ethereum(accountIndex),
        curve: import_openapi_wallet_service.DerivationInfoCurveEnum.secp256k1,
        algorithm: import_openapi_wallet_service.Algorithm.secp256k1,
        addressFormat: import_openapi_wallet_service.DerivationInfoAddressFormatEnum.ethereum
        // EVM chains use Ethereum address format
      };
    default:
      return null;
  }
}

// src/errors.ts
var import_axios = require("axios");
var WalletServiceError = class extends Error {
  constructor(data, errorName) {
    super(data.detail);
    this.name = errorName;
    this.type = data.type;
    this.title = data.title;
    this.detail = data.detail;
    this.requestId = data.requestId;
  }
};
function getAxiosErrorData(error) {
  if ((0, import_axios.isAxiosError)(error)) {
    return error.response?.data;
  }
  return void 0;
}
function getErrorMessage(error, fallbackMessage = "An error occurred") {
  if (error instanceof WalletServiceError) {
    return error.detail || error.title || error.message || fallbackMessage;
  }
  const data = getAxiosErrorData(error);
  if (data) {
    return data.message || data.detail || data.title || fallbackMessage;
  }
  if (error instanceof Error) {
    return error.message;
  }
  return fallbackMessage;
}
function isAuthenticationError(error) {
  if ((0, import_axios.isAxiosError)(error)) {
    return error.response?.status === 401 || error.response?.status === 403;
  }
  return false;
}
function parseWalletServiceError(data) {
  if (!data || !data.type) {
    return null;
  }
  switch (data.type) {
    case "spending-limit-exceeded":
      return new SpendingLimitError(data);
    case "transaction-blocked":
      return new TransactionBlockedError(data);
    default:
      return null;
  }
}
var SpendingLimitError = class extends WalletServiceError {
  constructor(data) {
    super(data, "SpendingLimitError");
    this.previousSpendCents = data.previousSpendCents;
    this.transactionSpendCents = data.transactionSpendCents;
    this.totalSpendCents = data.totalSpendCents;
    this.limitCents = data.limitCents;
  }
};
var TransactionBlockedError = class extends WalletServiceError {
  constructor(data) {
    super(data, "TransactionBlockedError");
    this.scannerResult = data.scannerResult;
  }
};

// src/PhantomClient.ts
var import_utils = require("@phantom/utils");
var PhantomClient = class {
  constructor(config, stamper) {
    this.config = {
      ...config,
      walletType: config.walletType || "user-wallet"
    };
    this.logger = config.logger;
    this.axiosInstance = import_axios2.default.create();
    const customHeaders = {};
    if (config.headers) {
      Object.assign(customHeaders, config.headers);
    }
    if (Object.keys(customHeaders).length > 0) {
      this.axiosInstance.interceptors.request.use((config2) => {
        Object.assign(config2.headers, customHeaders);
        return config2;
      });
    }
    this.axiosInstance.interceptors.response.use(
      (response) => {
        this.logger?.debug(
          `API response: method=${response.config.method?.toUpperCase()} url=${response.config.url} status=${response.status} body=${JSON.stringify(response.data)}`
        );
        return response;
      },
      (error) => {
        const status = error.response?.status;
        const url = error.config?.url;
        const method = error.config?.method?.toUpperCase();
        this.logger?.error(
          `API error response: method=${method} url=${url} status=${status} body=${JSON.stringify(error.response?.data)}`
        );
        return Promise.reject(error);
      }
    );
    if (stamper) {
      this.axiosInstance.interceptors.request.use(async (config2) => {
        return await this.stampRequest(config2, stamper);
      });
      this.stamper = stamper;
    }
    const configuration = new import_openapi_wallet_service2.Configuration({
      basePath: config.apiBaseUrl
    });
    this.kmsApi = new import_openapi_wallet_service2.KMSRPCApi(configuration, config.apiBaseUrl, this.axiosInstance);
  }
  setOrganizationId(organizationId) {
    if (!organizationId) {
      throw new Error("organizationId is required");
    }
    this.config.organizationId = organizationId;
  }
  /**
   * Gets the authenticator public key from the stamper if available.
   * Returns undefined if the stamper doesn't support key info retrieval.
   *
   * OIDC/Auth2 stampers (device-code flow) are excluded: their P-256 key is
   * not a KMS Keypair authenticator — it is only used for OIDC stamp signing —
   * so sending it as authenticatorPublicKey would cause a 400 (wrong key size).
   */
  getAuthenticatorPublicKey() {
    if (!this.stamper) {
      return void 0;
    }
    if (this.stamper.type === "OIDC") {
      return void 0;
    }
    if ("getKeyInfo" in this.stamper && typeof this.stamper.getKeyInfo === "function") {
      const keyInfo = this.stamper.getKeyInfo();
      return keyInfo?.publicKey;
    }
    return void 0;
  }
  async createWallet(walletName) {
    try {
      if (!this.config.organizationId) {
        throw new Error("organizationId is required to create a wallet");
      }
      const timestamp = await (0, import_utils.getSecureTimestamp)();
      const walletRequest = {
        organizationId: this.config.organizationId,
        walletName: walletName || `Wallet ${timestamp}`,
        accounts: [
          DerivationPath.Solana(),
          DerivationPath.Ethereum(),
          DerivationPath.Bitcoin(),
          DerivationPath.Sui()
        ]
      };
      const request = {
        method: import_openapi_wallet_service2.CreateWalletMethodEnum.createWallet,
        params: walletRequest,
        timestampMs: timestamp
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const walletResult = response.data.result;
      const requestAccounts = {
        method: import_openapi_wallet_service2.GetAccountsMethodEnum.getAccounts,
        params: {
          accounts: [
            DerivationPath.Solana(),
            DerivationPath.Ethereum(),
            DerivationPath.Bitcoin(),
            DerivationPath.Sui()
          ],
          organizationId: this.config.organizationId,
          walletId: walletResult.walletId
        },
        timestampMs: timestamp
      };
      const accountsResponse = await this.kmsApi.postKmsRpc(requestAccounts);
      const accountsResult = accountsResponse.data.result;
      return {
        walletId: walletResult.walletId,
        addresses: accountsResult.map((account) => ({
          addressType: account.addressFormat,
          address: account.address
        }))
      };
    } catch (error) {
      this.logger?.error(`Failed to create wallet: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to create wallet: ${error.response?.data?.message || error.message}`);
    }
  }
  async prepare(transaction, organizationId, submissionConfig, account, authenticatorPublicKey, methodName) {
    try {
      const request = {
        transaction,
        organizationId,
        submissionConfig,
        simulationConfig: { account },
        authenticatorPublicKey
      };
      this.logger?.debug(
        `prepare request: submissionConfig=${JSON.stringify(submissionConfig)} account=${account} methodName=${methodName} txLength=${transaction.length} request=${JSON.stringify(request)}`
      );
      const response = await this.axiosInstance.post(`${this.config.apiBaseUrl}/prepare`, request, {
        headers: {
          "Content-Type": "application/json",
          "X-Rpc-Method": methodName
        }
      });
      return response.data;
    } catch (error) {
      const errorData = getAxiosErrorData(error);
      const axiosStatus = error.response?.status;
      const rawResponseData = error.response?.data;
      this.logger?.error(
        `[TX_DEBUG] prepare failed: status=${axiosStatus} methodName=${methodName} body=${JSON.stringify(rawResponseData)}`
      );
      const walletServiceError = parseWalletServiceError(errorData);
      if (walletServiceError) {
        throw walletServiceError;
      }
      const message = errorData?.detail || getErrorMessage(error, "Failed to submit transaction");
      throw new Error(message);
    }
  }
  async getTransactionForSigning(params) {
    const { encodedTransaction, networkId, submissionConfig, account, authenticatorPublicKey } = params;
    const isEvmTransaction = (0, import_utils.isEthereumChain)(networkId);
    const isSolanaTransaction = (0, import_utils.isSolanaChain)(networkId);
    if (isEvmTransaction) {
      return { kind: "RLP_ENCODED", bytes: encodedTransaction };
    }
    if (isSolanaTransaction && this.config.walletType === "user-wallet") {
      if (!account) {
        throw new Error("Account is required to simulate Solana transactions with spending limits");
      }
      try {
        const prepareResponse = await this.prepare(
          encodedTransaction,
          this.config.organizationId,
          submissionConfig,
          account,
          authenticatorPublicKey,
          params.methodName
        );
        return prepareResponse.transaction;
      } catch (e) {
        if (e instanceof WalletServiceError) {
          throw e;
        }
        const errorMessage = e instanceof Error ? e.message : String(e);
        throw new Error(errorMessage);
      }
    }
    return encodedTransaction;
  }
  getRpcMethodName(networkId, includeSubmissionConfig) {
    const isEvm = (0, import_utils.isEthereumChain)(networkId);
    if (isEvm) {
      return includeSubmissionConfig ? "eth_sendTransaction" : "eth_signTransaction";
    }
    return includeSubmissionConfig ? "signAndSendTransaction" : "signTransaction";
  }
  /**
   * Private method for shared signing logic
   */
  async performTransactionSigning(params, includeSubmissionConfig, presignTransaction) {
    const walletId = params.walletId;
    const encodedTransaction = params.transaction;
    const networkIdParam = params.networkId;
    const derivationIndex = params.derivationIndex ?? 0;
    const methodName = this.getRpcMethodName(networkIdParam, includeSubmissionConfig);
    try {
      if (!this.config.organizationId) {
        throw new Error("organizationId is required to sign a transaction");
      }
      const submissionConfig = deriveSubmissionConfig(networkIdParam);
      if (!submissionConfig) {
        throw new Error(`SubmissionConfig could not be derived for network ID: ${networkIdParam}`);
      }
      const networkConfig = getNetworkConfig(networkIdParam, derivationIndex);
      if (!networkConfig) {
        throw new Error(`Unsupported network ID: ${networkIdParam}`);
      }
      const derivationInfo = {
        derivationPath: networkConfig.derivationPath,
        curve: networkConfig.curve,
        addressFormat: networkConfig.addressFormat
      };
      const authenticatorPublicKey = this.getAuthenticatorPublicKey();
      let transactionForSigning = await this.getTransactionForSigning({
        encodedTransaction,
        networkId: networkIdParam,
        submissionConfig,
        authenticatorPublicKey,
        account: params.account,
        methodName
      });
      if (presignTransaction && typeof transactionForSigning === "string") {
        try {
          transactionForSigning = await presignTransaction(transactionForSigning, {
            networkId: networkIdParam,
            walletId
          });
        } catch (hookError) {
          throw new Error(`presignTransaction hook failed: ${getErrorMessage(hookError, "unknown error")}`);
        }
      }
      const signRequest = {
        organizationId: this.config.organizationId,
        walletId,
        transaction: transactionForSigning,
        derivationInfo
      };
      if (includeSubmissionConfig) {
        signRequest.submissionConfig = submissionConfig;
      }
      if (includeSubmissionConfig && params.account) {
        signRequest.simulationConfig = {
          account: params.account
        };
      }
      const request = {
        method: import_openapi_wallet_service2.SignTransactionMethodEnum.signTransaction,
        params: signRequest,
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      this.logger?.debug(
        `[TX_DEBUG] KMS request: methodName=${methodName} networkId=${networkIdParam} walletId=${walletId} account=${params.account}`
      );
      this.logger?.debug(`[TX_DEBUG] transactionForSigning=${JSON.stringify(transactionForSigning)}`);
      const response = await this.kmsApi.postKmsRpc(request, {
        headers: {
          "X-Rpc-Method": methodName
        }
      });
      const result = response.data.result;
      const rpcSubmissionResult = response.data["rpc_submission_result"];
      this.logger?.debug(
        `[TX_DEBUG] KMS response: methodName=${methodName} hasResult=${!!result} rpcSubmissionResult=${JSON.stringify(rpcSubmissionResult)} fullResponse=${JSON.stringify(response.data)}`
      );
      let hash = void 0;
      if (includeSubmissionConfig && rpcSubmissionResult) {
        if (rpcSubmissionResult.error) {
          const { error } = rpcSubmissionResult;
          throw new Error(
            `Transaction broadcast failed: ${error.data ?? error.message ?? JSON.stringify(rpcSubmissionResult)}`
          );
        }
        const envelope = rpcSubmissionResult.result;
        if (envelope != null) {
          if (typeof envelope === "string") {
            hash = envelope;
          } else if (typeof envelope.result === "string") {
            hash = envelope.result;
          } else if (envelope.error) {
            const { error } = envelope;
            throw new Error(`Transaction broadcast failed: ${error.data ?? error.message ?? JSON.stringify(envelope)}`);
          }
        }
      }
      return {
        signedTransaction: result.transaction,
        // Base64 encoded signed transaction
        hash
      };
    } catch (error) {
      const actionType = includeSubmissionConfig ? "sign and send" : "sign";
      if (error instanceof WalletServiceError) {
        throw error;
      }
      const axiosStatus = error.response?.status;
      const axiosData = error.response?.data;
      this.logger?.error(
        `[TX_DEBUG] performTransactionSigning failed: methodName=${methodName} status=${axiosStatus} body=${JSON.stringify(axiosData)}`
      );
      throw new Error(getErrorMessage(error, `Failed to ${actionType} transaction`));
    }
  }
  /**
   * Sign a transaction
   */
  async signTransaction(params) {
    const result = await this.performTransactionSigning(params, false, params.presignTransaction);
    return {
      rawTransaction: result.signedTransaction
    };
  }
  /**
   * Sign and send a transaction
   */
  async signAndSendTransaction(params) {
    const result = await this.performTransactionSigning(params, true, params.presignTransaction);
    return {
      rawTransaction: result.signedTransaction,
      hash: result.hash
    };
  }
  async getWalletAddresses(walletId, derivationPaths, derivationIndex) {
    try {
      const accountIndex = derivationIndex ?? 0;
      const paths = derivationPaths || [
        DerivationPath.Solana(accountIndex),
        DerivationPath.Ethereum(accountIndex),
        DerivationPath.Bitcoin(accountIndex),
        DerivationPath.Sui(accountIndex)
      ];
      const requestAccounts = {
        method: import_openapi_wallet_service2.GetAccountsMethodEnum.getAccounts,
        params: {
          accounts: paths,
          organizationId: this.config.organizationId,
          walletId
        },
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const accountsResponse = await this.kmsApi.postKmsRpc(requestAccounts);
      const accountsResult = accountsResponse.data.result;
      const addresses = accountsResult.map((account) => ({
        addressType: account.addressFormat,
        address: account.address
      }));
      return addresses;
    } catch (error) {
      this.logger?.error(`Failed to get wallet addresses: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to get wallet addresses: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Sign an Ethereum message using EIP-191 personal sign
   */
  async ethereumSignMessage(params) {
    const walletId = params.walletId;
    const messageParam = params.message;
    const networkIdParam = params.networkId;
    const derivationIndex = params.derivationIndex ?? 0;
    try {
      if (!this.config.organizationId) {
        throw new Error("organizationId is required to sign a message");
      }
      const networkConfig = getNetworkConfig(networkIdParam, derivationIndex);
      if (!networkConfig) {
        throw new Error(`Unsupported network ID: ${networkIdParam}`);
      }
      const derivationInfo = {
        derivationPath: networkConfig.derivationPath,
        curve: networkConfig.curve,
        addressFormat: networkConfig.addressFormat
      };
      const base64StringMessage = messageParam;
      const request = {
        method: "ethereumSignMessage",
        params: {
          message: base64StringMessage,
          organizationId: this.config.organizationId,
          walletId,
          derivationInfo
        },
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result.signature;
    } catch (error) {
      this.logger?.error(`Failed to sign Ethereum message: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to sign Ethereum message: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Sign a UTF-8 message for Solana
   */
  async signUtf8Message(params) {
    const walletId = params.walletId;
    const messageParam = params.message;
    const networkIdParam = params.networkId;
    const derivationIndex = params.derivationIndex ?? 0;
    try {
      if (!this.config.organizationId) {
        throw new Error("organizationId is required to sign a message");
      }
      const networkConfig = getNetworkConfig(networkIdParam, derivationIndex);
      if (!networkConfig) {
        throw new Error(`Unsupported network ID: ${networkIdParam}`);
      }
      const derivationInfo = {
        derivationPath: networkConfig.derivationPath,
        curve: networkConfig.curve,
        addressFormat: networkConfig.addressFormat
      };
      const signRequest = {
        organizationId: this.config.organizationId,
        walletId,
        message: messageParam,
        algorithm: networkConfig.algorithm,
        derivationInfo
      };
      const request = {
        method: import_openapi_wallet_service2.SignUTF8MessageMethodEnum.signUtf8Message,
        params: signRequest,
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result.signature;
    } catch (error) {
      this.logger?.error(`Failed to sign raw payload: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to sign raw payload: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Sign EIP-712 typed data for Ethereum
   */
  async ethereumSignTypedData(params) {
    const walletId = params.walletId;
    const typedData = params.typedData;
    const networkIdParam = params.networkId;
    const derivationIndex = params.derivationIndex ?? 0;
    try {
      if (!this.config.organizationId) {
        throw new Error("organizationId is required to sign typed data");
      }
      const networkConfig = getNetworkConfig(networkIdParam, derivationIndex);
      if (!networkConfig) {
        throw new Error(`Unsupported network ID: ${networkIdParam}`);
      }
      const derivationInfo = {
        derivationPath: networkConfig.derivationPath,
        curve: networkConfig.curve,
        addressFormat: networkConfig.addressFormat
      };
      const request = {
        method: "ethereumSignTypedData",
        params: {
          typedData,
          organizationId: this.config.organizationId,
          walletId,
          derivationInfo
        },
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result.signature;
    } catch (error) {
      this.logger?.error(`Failed to sign typed data: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to sign typed data: ${error.response?.data?.message || error.message}`);
    }
  }
  async getWallets(limit, offset) {
    try {
      const request = {
        method: "getOrganizationWallets",
        params: {
          organizationId: this.config.organizationId,
          limit: limit || 20,
          offset: offset || 0
        },
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return {
        wallets: result.wallets.map((wallet) => ({
          walletId: wallet.walletId,
          walletName: wallet.walletName
        })),
        totalCount: result.totalCount,
        limit: result.limit,
        offset: result.offset
      };
    } catch (error) {
      this.logger?.error(`Failed to get wallets: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to get wallets: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Get organization details by organization ID
   */
  async getOrganization(organizationId) {
    try {
      const request = {
        method: "getOrganization",
        params: {
          organizationId
        },
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result;
    } catch (error) {
      this.logger?.error(`Failed to get organization: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to get organization: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Create a new organization with the specified name and users
   * @param name Organization name
   * @param users Array of users with their authenticators
   */
  validateNameLength(name, type) {
    const MAX_NAME_LENGTH = 64;
    if (name.length > MAX_NAME_LENGTH) {
      throw new Error(`${type} name cannot exceed ${MAX_NAME_LENGTH} characters. Current length: ${name.length}`);
    }
  }
  async createOrganization(name, users, tags) {
    try {
      if (!name) {
        throw new Error("Organization name is required");
      }
      this.validateNameLength(name, "Organization");
      if (!users || users.length === 0) {
        throw new Error("At least one user is required");
      }
      for (const user of users) {
        if (user.username) {
          this.validateNameLength(user.username, "Username");
        }
        for (const auth of user.authenticators) {
          if (auth.authenticatorName) {
            this.validateNameLength(auth.authenticatorName, "Authenticator");
          }
        }
      }
      const params = {
        organizationName: name,
        users: users.map((userConfig) => ({
          username: userConfig.username || `user-${(0, import_utils.randomUUID)()}`,
          authenticators: userConfig.authenticators,
          policy: userConfig.role === "ADMIN" ? {
            type: import_openapi_wallet_service2.UserPolicyOneOfTypeEnum.root
          } : { type: "CEL", preset: "LEGACY_USER_ROLE" }
        })),
        tags
      };
      const request = {
        method: import_openapi_wallet_service2.CreateOrganizationMethodEnum.createOrganization,
        params,
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result;
    } catch (error) {
      this.logger?.error(`Failed to create organization: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to create organization: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Create an authenticator for a user in an organization
   */
  async createAuthenticator(params) {
    try {
      if (params.username) {
        this.validateNameLength(params.username, "Username");
      }
      if (params.authenticatorName) {
        this.validateNameLength(params.authenticatorName, "Authenticator");
      }
      if (params.authenticator?.authenticatorName) {
        this.validateNameLength(params.authenticator.authenticatorName, "Authenticator");
      }
      const requestParams = {
        organizationId: params.organizationId,
        username: params.username,
        authenticatorName: params.authenticatorName,
        authenticator: params.authenticator,
        replaceExpirable: params.replaceExpirable
      };
      const request = {
        method: import_openapi_wallet_service2.CreateAuthenticatorMethodEnum.createAuthenticator,
        params: requestParams,
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result;
    } catch (error) {
      this.logger?.error(`Failed to create authenticator: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to create authenticator: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Delete an authenticator for a user in an organization
   */
  async deleteAuthenticator(params) {
    try {
      const requestParams = {
        organizationId: params.organizationId,
        username: params.username,
        authenticatorId: params.authenticatorId
      };
      const request = {
        method: import_openapi_wallet_service2.DeleteAuthenticatorMethodEnum.deleteAuthenticator,
        params: requestParams,
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result;
    } catch (error) {
      this.logger?.error(`Failed to delete authenticator: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to delete authenticator: ${error.response?.data?.message || error.message}`);
    }
  }
  async grantOrganizationAccess(params) {
    try {
      const request = {
        method: import_openapi_wallet_service2.GrantOrganizationAccessMethodEnum.grantOrganizationAccess,
        params,
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result;
    } catch (error) {
      this.logger?.error(
        `Failed to grant organization access: ${JSON.stringify(error.response?.data) || error.message}`
      );
      throw new Error(`Failed to grant organization access: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Add a new user to an organization
   */
  async addUserToOrganization(params) {
    try {
      const request = {
        method: import_openapi_wallet_service2.AddUserToOrganizationMethodEnum.addUserToOrganization,
        params,
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      await this.kmsApi.postKmsRpc(request);
    } catch (error) {
      this.logger?.error(
        `Failed to add user to organization: ${JSON.stringify(error.response?.data) || error.message}`
      );
      throw new Error(`Failed to add user to organization: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Get a wallet by tag from the specified organization
   */
  async getWalletWithTag(params) {
    try {
      const request = {
        method: "getWalletWithTag",
        params: {
          organizationId: params.organizationId,
          tag: params.tag,
          derivationPaths: params.derivationPaths
        },
        timestampMs: await (0, import_utils.getSecureTimestamp)()
      };
      const response = await this.kmsApi.postKmsRpc(request);
      const result = response.data.result;
      return result;
    } catch (error) {
      this.logger?.error(`Failed to get wallet with tag: ${JSON.stringify(error.response?.data) || error.message}`);
      throw new Error(`Failed to get wallet with tag: ${error.response?.data?.message || error.message}`);
    }
  }
  /**
   * Stamp an axios request with the provided stamper
   */
  async stampRequest(config, stamper) {
    const requestBody = typeof config.data === "string" ? config.data : config.data === void 0 ? "" : JSON.stringify(config.data);
    const dataUtf8 = import_buffer.Buffer.from(requestBody, "utf8");
    const stamp = await stamper.stamp({
      data: dataUtf8
    });
    config.headers = config.headers || {};
    config.headers["X-Phantom-Stamp"] = stamp;
    const dynamicHeaders = this.config.getHeaders?.();
    if (dynamicHeaders) {
      for (const [headerName, headerValue] of Object.entries(dynamicHeaders)) {
        if (typeof headerValue === "string" && headerValue) {
          config.headers[headerName] = headerValue;
        }
      }
    }
    return config;
  }
};

// src/index.ts
var import_crypto = require("@phantom/crypto");

// src/types.ts
var import_utils2 = require("@phantom/utils");

// src/index.ts
var import_constants4 = require("@phantom/constants");
var import_openapi_wallet_service3 = require("@phantom/openapi-wallet-service");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AddressType,
  DerivationPath,
  NetworkId,
  PhantomClient,
  SpendingLimitError,
  TransactionBlockedError,
  WalletServiceError,
  deriveSubmissionConfig,
  generateKeyPair,
  getAxiosErrorData,
  getDerivationPathForNetwork,
  getErrorMessage,
  getNetworkConfig,
  getNetworkDescription,
  getNetworkIdsByChain,
  getSupportedNetworkIds,
  isAuthenticationError,
  parseWalletServiceError,
  supportsTransactionSubmission
});
