// @phantom/utils@2.0.1/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  getChainPrefix: () => getChainPrefix,
  getSecureTimestamp: () => getSecureTimestamp,
  getSecureTimestampSync: () => getSecureTimestampSync,
  isEthereumChain: () => isEthereumChain,
  isSolanaChain: () => isSolanaChain,
  randomString: () => randomString,
  randomUUID: () => randomUUID
});
module.exports = __toCommonJS(src_exports);

// src/uuid.ts
function randomUUID() {
  if (typeof crypto !== "undefined" && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(c) {
    const r = Math.random() * 16 | 0;
    const v = c === "x" ? r : r & 3 | 8;
    return v.toString(16);
  });
}
function randomString(length) {
  const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  let result = "";
  for (let i = 0; i < length; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

// src/time.ts
var TimeService = class {
  constructor() {
    this.cache = null;
    this.CACHE_DURATION = 3e4;
    // 30 seconds cache
    this.TIME_API_URL = "https://time.phantom.app/utc";
  }
  static getInstance() {
    if (!TimeService.instance) {
      TimeService.instance = new TimeService();
    }
    return TimeService.instance;
  }
  /**
   * Get current timestamp from Phantom's secure time API
   * Includes basic caching to reduce API calls
   */
  async now() {
    const now = Date.now();
    if (this.cache && now - this.cache.fetchedAt < this.CACHE_DURATION) {
      const elapsed = now - this.cache.fetchedAt;
      return this.cache.timestamp + elapsed;
    }
    try {
      const response = await fetch(this.TIME_API_URL);
      if (!response.ok) {
        throw new Error(`Time API responded with status: ${response.status}`);
      }
      const timestampText = await response.text();
      const timestamp = parseInt(timestampText, 10);
      if (isNaN(timestamp)) {
        throw new Error(`Invalid timestamp received: ${timestampText}`);
      }
      this.cache = {
        timestamp,
        fetchedAt: now
      };
      return timestamp;
    } catch (error) {
      return Date.now();
    }
  }
  /**
   * Synchronous version that uses cached time if available,
   * otherwise falls back to Date.now()
   */
  nowSync() {
    if (this.cache) {
      const elapsed = Date.now() - this.cache.fetchedAt;
      if (elapsed < this.CACHE_DURATION) {
        return this.cache.timestamp + elapsed;
      }
    }
    return Date.now();
  }
  /**
   * Clear the cache (useful for testing)
   */
  clearCache() {
    this.cache = null;
  }
};
var timeService = TimeService.getInstance();
var getSecureTimestamp = () => timeService.now();
var getSecureTimestampSync = () => timeService.nowSync();

// src/network.ts
function getChainPrefix(networkId) {
  return networkId.split(":")[0].toLowerCase();
}
function isEthereumChain(networkId) {
  const network = getChainPrefix(networkId);
  return network === "eip155";
}
function isSolanaChain(networkId) {
  const network = getChainPrefix(networkId);
  return network === "solana";
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getChainPrefix,
  getSecureTimestamp,
  getSecureTimestampSync,
  isEthereumChain,
  isSolanaChain,
  randomString,
  randomUUID
});
