// @phantom/openapi-wallet-service@0.1.11/model/add-user-to-organization-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { PartialKmsUser } from "./partial-kms-user";

/**
 *
 * @export
 * @interface AddUserToOrganizationRequest
 */
export interface AddUserToOrganizationRequest {
  /**
   *
   * @type {string}
   * @memberof AddUserToOrganizationRequest
   */
  organizationId: string;
  /**
   * If true, replace the oldest expirable user if one exists and reaching the user limit per organization. If none exists, reaching the user limit per organization will disallow adding more users.
   * @type {boolean}
   * @memberof AddUserToOrganizationRequest
   */
  replaceExpirable?: boolean;
  /**
   *
   * @type {PartialKmsUser}
   * @memberof AddUserToOrganizationRequest
   */
  user: PartialKmsUser;
}
