// @phantom/openapi-wallet-service@0.1.11/model/cel-policy-preset-one-of2.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { UsdLimitSettings } from "./usd-limit-settings";

/**
 * A preset for user policy created for a specific dapp connection (\"User Wallet\" use case).
 * @export
 * @interface CelPolicyPresetOneOf2
 */
export interface CelPolicyPresetOneOf2 {
  /**
   *
   * @type {string}
   * @memberof CelPolicyPresetOneOf2
   */
  preset: CelPolicyPresetOneOf2PresetEnum;
  /**
   *
   * @type {UsdLimitSettings}
   * @memberof CelPolicyPresetOneOf2
   */
  usdLimit?: UsdLimitSettings | null;
  /**
   *
   * @type {string}
   * @memberof CelPolicyPresetOneOf2
   */
  walletId: string;
}

/**
 * @export
 * @enum {string}
 */
export enum CelPolicyPresetOneOf2PresetEnum {
  dappConnectionUser = "DAPP_CONNECTION_USER",
}
