// @phantom/openapi-wallet-service@0.1.11/model/complete-import-wallet.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { CompleteImportWalletRequest } from "./complete-import-wallet-request";

/**
 *
 * @export
 * @interface CompleteImportWallet
 */
export interface CompleteImportWallet {
  /**
   *
   * @type {string}
   * @memberof CompleteImportWallet
   */
  method: CompleteImportWalletMethodEnum;
  /**
   *
   * @type {CompleteImportWalletRequest}
   * @memberof CompleteImportWallet
   */
  params: CompleteImportWalletRequest;
}

/**
 * @export
 * @enum {string}
 */
export enum CompleteImportWalletMethodEnum {
  completeImportWallet = "completeImportWallet",
}
