// @phantom/openapi-wallet-service@0.1.11/model/create-authenticator-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { PartialKmsAuthenticator } from "./partial-kms-authenticator";

/**
 *
 * @export
 * @interface CreateAuthenticatorRequest
 */
export interface CreateAuthenticatorRequest {
  /**
   *
   * @type {PartialKmsAuthenticator}
   * @memberof CreateAuthenticatorRequest
   */
  authenticator: PartialKmsAuthenticator;
  /**
   *
   * @type {string}
   * @memberof CreateAuthenticatorRequest
   */
  organizationId: string;
  /**
   * Replace the oldest expirable authenticator if one exists, and the user has reached the limit of allowed authenticators. If none exists, reaching the authenticator limit per user will disallow adding more authenticators.
   * @type {boolean}
   * @memberof CreateAuthenticatorRequest
   */
  replaceExpirable?: boolean;
  /**
   *
   * @type {string}
   * @memberof CreateAuthenticatorRequest
   */
  username: string;
}
