// @phantom/openapi-wallet-service@0.1.11/model/delete-authenticator.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { DeleteAuthenticatorRequest } from "./delete-authenticator-request";

/**
 *
 * @export
 * @interface DeleteAuthenticator
 */
export interface DeleteAuthenticator {
  /**
   *
   * @type {string}
   * @memberof DeleteAuthenticator
   */
  method: DeleteAuthenticatorMethodEnum;
  /**
   *
   * @type {DeleteAuthenticatorRequest}
   * @memberof DeleteAuthenticator
   */
  params: DeleteAuthenticatorRequest;
}

/**
 * @export
 * @enum {string}
 */
export enum DeleteAuthenticatorMethodEnum {
  deleteAuthenticator = "deleteAuthenticator",
}
