// @phantom/openapi-wallet-service@0.1.11/model/delete-user-from-organization.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { DeleteUserFromOrganizationRequest } from "./delete-user-from-organization-request";

/**
 *
 * @export
 * @interface DeleteUserFromOrganization
 */
export interface DeleteUserFromOrganization {
  /**
   *
   * @type {string}
   * @memberof DeleteUserFromOrganization
   */
  method: DeleteUserFromOrganizationMethodEnum;
  /**
   *
   * @type {DeleteUserFromOrganizationRequest}
   * @memberof DeleteUserFromOrganization
   */
  params: DeleteUserFromOrganizationRequest;
}

/**
 * @export
 * @enum {string}
 */
export enum DeleteUserFromOrganizationMethodEnum {
  deleteUserFromOrganization = "deleteUserFromOrganization",
}
