// @phantom/openapi-wallet-service@0.1.11/model/ethereum-sign-typed-data-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { DerivationInfoSchema } from "./derivation-info-schema";
// May contain unused imports in some cases
// @ts-ignore
import type { EthereumTypedData } from "./ethereum-typed-data";

/**
 *
 * @export
 * @interface EthereumSignTypedDataRequest
 */
export interface EthereumSignTypedDataRequest {
  /**
   *
   * @type {DerivationInfoSchema}
   * @memberof EthereumSignTypedDataRequest
   */
  derivationInfo?: DerivationInfoSchema | null;
  /**
   *
   * @type {string}
   * @memberof EthereumSignTypedDataRequest
   */
  organizationId: string;
  /**
   *
   * @type {string}
   * @memberof EthereumSignTypedDataRequest
   */
  policyName?: string | null;
  /**
   *
   * @type {EthereumTypedData}
   * @memberof EthereumSignTypedDataRequest
   */
  typedData: EthereumTypedData;
  /**
   *
   * @type {string}
   * @memberof EthereumSignTypedDataRequest
   */
  walletId: string;
}
