// @phantom/openapi-wallet-service@0.1.11/model/external-kms-organization.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { KmsOrgUser } from "./kms-org-user";

/**
 *
 * @export
 * @interface ExternalKmsOrganization
 */
export interface ExternalKmsOrganization {
  /**
   *
   * @type {string}
   * @memberof ExternalKmsOrganization
   */
  organizationId: string;
  /**
   *
   * @type {string}
   * @memberof ExternalKmsOrganization
   */
  organizationName: string;
  /**
   *
   * @type {string}
   * @memberof ExternalKmsOrganization
   */
  parentOrganizationId?: string | null;
  /**
   *
   * @type {Array<string>}
   * @memberof ExternalKmsOrganization
   */
  tags: Array<string>;
  /**
   *
   * @type {Array<KmsOrgUser>}
   * @memberof ExternalKmsOrganization
   */
  users: Array<KmsOrgUser>;
}
