// @phantom/openapi-wallet-service@0.1.11/model/get-wallet-access-grants-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 *
 * @export
 * @interface GetWalletAccessGrantsRequest
 */
export interface GetWalletAccessGrantsRequest {
  /**
   *
   * @type {number}
   * @memberof GetWalletAccessGrantsRequest
   */
  limit?: number;
  /**
   *
   * @type {number}
   * @memberof GetWalletAccessGrantsRequest
   */
  offset?: number;
  /**
   *
   * @type {string}
   * @memberof GetWalletAccessGrantsRequest
   */
  ownerOrganizationId: string;
  /**
   *
   * @type {string}
   * @memberof GetWalletAccessGrantsRequest
   */
  targetOrganizationId: string;
  /**
   *
   * @type {string}
   * @memberof GetWalletAccessGrantsRequest
   */
  walletId: string;
}
