// @phantom/openapi-wallet-service@0.1.11/model/get-wallet-access-grants.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { GetWalletAccessGrantsRequest } from "./get-wallet-access-grants-request";

/**
 *
 * @export
 * @interface GetWalletAccessGrants
 */
export interface GetWalletAccessGrants {
  /**
   *
   * @type {string}
   * @memberof GetWalletAccessGrants
   */
  method: GetWalletAccessGrantsMethodEnum;
  /**
   *
   * @type {GetWalletAccessGrantsRequest}
   * @memberof GetWalletAccessGrants
   */
  params: GetWalletAccessGrantsRequest;
}

/**
 * @export
 * @enum {string}
 */
export enum GetWalletAccessGrantsMethodEnum {
  getWalletAccessGrantsRequest = "getWalletAccessGrantsRequest",
}
