// @phantom/openapi-wallet-service@0.1.11/model/grant-organization-access-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { VersionedOrganizationPolicy } from "./versioned-organization-policy";
// May contain unused imports in some cases
// @ts-ignore
import type { VersionedTransactionPolicy } from "./versioned-transaction-policy";

/**
 *
 * @export
 * @interface GrantOrganizationAccessRequest
 */
export interface GrantOrganizationAccessRequest {
  /**
   *
   * @type {string}
   * @memberof GrantOrganizationAccessRequest
   */
  ownerOrganizationId: string;
  /**
   *
   * @type {VersionedOrganizationPolicy}
   * @memberof GrantOrganizationAccessRequest
   */
  policy: VersionedOrganizationPolicy;
  /**
   * If specified, this grant be added in addition to another grant(s) to the same target organization for the same wallet (i.e. with different policies). However, access through this grant, will only be possible if policy name is specified in the request.
   * @type {string}
   * @memberof GrantOrganizationAccessRequest
   */
  policyName?: string | null;
  /**
   *
   * @type {string}
   * @memberof GrantOrganizationAccessRequest
   */
  targetOrganizationId: string;
  /**
   *
   * @type {VersionedTransactionPolicy}
   * @memberof GrantOrganizationAccessRequest
   */
  transactionPolicy?: VersionedTransactionPolicy | null;
  /**
   *
   * @type {string}
   * @memberof GrantOrganizationAccessRequest
   */
  walletId: string;
}
