// @phantom/openapi-wallet-service@0.1.11/model/grant-organization-access.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { GrantOrganizationAccessRequest } from "./grant-organization-access-request";

/**
 * Share access to a wallet with another organization.
 * @export
 * @interface GrantOrganizationAccess
 */
export interface GrantOrganizationAccess {
  /**
   *
   * @type {string}
   * @memberof GrantOrganizationAccess
   */
  method: GrantOrganizationAccessMethodEnum;
  /**
   * Share access to a wallet with another organization.
   * @type {GrantOrganizationAccessRequest}
   * @memberof GrantOrganizationAccess
   */
  params: GrantOrganizationAccessRequest;
}

/**
 * @export
 * @enum {string}
 */
export enum GrantOrganizationAccessMethodEnum {
  grantOrganizationAccess = "grantOrganizationAccess",
}
