// @phantom/openapi-wallet-service@0.1.11/model/import-mnemonic-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { DerivationInfo } from "./derivation-info";
// May contain unused imports in some cases
// @ts-ignore
import type { ImportEnvelope } from "./import-envelope";

/**
 *
 * @export
 * @interface ImportMnemonicRequest
 */
export interface ImportMnemonicRequest {
  /**
   *
   * @type {Array<DerivationInfo>}
   * @memberof ImportMnemonicRequest
   */
  accounts: Array<DerivationInfo>;
  /**
   *
   * @type {Array<number>}
   * @memberof ImportMnemonicRequest
   */
  encryptedEntropy: Array<number>;
  /**
   *
   * @type {ImportEnvelope}
   * @memberof ImportMnemonicRequest
   */
  importEnvelope: ImportEnvelope;
  /**
   *
   * @type {number}
   * @memberof ImportMnemonicRequest
   */
  mnemonicLength?: number;
  /**
   *
   * @type {string}
   * @memberof ImportMnemonicRequest
   */
  organizationId: string;
  /**
   *
   * @type {Array<string>}
   * @memberof ImportMnemonicRequest
   */
  tags?: Array<string>;
  /**
   *
   * @type {string}
   * @memberof ImportMnemonicRequest
   */
  walletName: string;
}
