// @phantom/openapi-wallet-service@0.1.11/model/instruction-validation-rule-one-of.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 *
 * @export
 * @interface InstructionValidationRuleOneOf
 */
export interface InstructionValidationRuleOneOf {
  /**
   *
   * @type {number}
   * @memberof InstructionValidationRuleOneOf
   */
  max?: number | null;
  /**
   *
   * @type {number}
   * @memberof InstructionValidationRuleOneOf
   */
  min?: number | null;
  /**
   *
   * @type {string}
   * @memberof InstructionValidationRuleOneOf
   */
  rule: InstructionValidationRuleOneOfRuleEnum;
}

/**
 * @export
 * @enum {string}
 */
export enum InstructionValidationRuleOneOfRuleEnum {
  systemInstructionLamportsThreshold = "SystemInstructionLamportsThreshold",
}
