// @phantom/openapi-wallet-service@0.1.11/model/instruction-validation-rule-one-of2.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 *
 * @export
 * @interface InstructionValidationRuleOneOf2
 */
export interface InstructionValidationRuleOneOf2 {
  /**
   *
   * @type {number}
   * @memberof InstructionValidationRuleOneOf2
   */
  max?: number | null;
  /**
   *
   * @type {number}
   * @memberof InstructionValidationRuleOneOf2
   */
  min?: number | null;
  /**
   *
   * @type {Array<string>}
   * @memberof InstructionValidationRuleOneOf2
   */
  mintAllowlist?: Array<string> | null;
  /**
   *
   * @type {Array<string>}
   * @memberof InstructionValidationRuleOneOf2
   */
  recipientAllowlist?: Array<string> | null;
  /**
   *
   * @type {string}
   * @memberof InstructionValidationRuleOneOf2
   */
  rule: InstructionValidationRuleOneOf2RuleEnum;
}

/**
 * @export
 * @enum {string}
 */
export enum InstructionValidationRuleOneOf2RuleEnum {
  transferCheckedConstraints = "TransferCheckedConstraints",
}
