// @phantom/openapi-wallet-service@0.1.11/model/kms-org-user.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { ExternalKmsAuthenticator } from "./external-kms-authenticator";
// May contain unused imports in some cases
// @ts-ignore
import type { UserPolicy } from "./user-policy";

/**
 *
 * @export
 * @interface KmsOrgUser
 */
export interface KmsOrgUser {
  /**
   *
   * @type {Array<ExternalKmsAuthenticator>}
   * @memberof KmsOrgUser
   */
  authenticators: Array<ExternalKmsAuthenticator>;
  /**
   *
   * @type {number}
   * @memberof KmsOrgUser
   */
  createdAtMs?: number | null;
  /**
   *
   * @type {number}
   * @memberof KmsOrgUser
   */
  expiresAtMs?: number | null;
  /**
   *
   * @type {UserPolicy}
   * @memberof KmsOrgUser
   */
  policy: UserPolicy;
  /**
   *
   * @type {{ [key: string]: string; }}
   * @memberof KmsOrgUser
   */
  traits?: { [key: string]: string };
  /**
   *
   * @type {string}
   * @memberof KmsOrgUser
   */
  username: string;
}
