// @phantom/openapi-wallet-service@0.1.11/model/organization-to-wallet.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { VersionedOrganizationPolicy } from "./versioned-organization-policy";
// May contain unused imports in some cases
// @ts-ignore
import type { VersionedTransactionPolicy } from "./versioned-transaction-policy";

/**
 *
 * @export
 * @interface OrganizationToWallet
 */
export interface OrganizationToWallet {
  /**
   *
   * @type {number}
   * @memberof OrganizationToWallet
   */
  deletedAtMs?: number | null;
  /**
   *
   * @type {string}
   * @memberof OrganizationToWallet
   */
  organizationId: string;
  /**
   *
   * @type {VersionedOrganizationPolicy}
   * @memberof OrganizationToWallet
   */
  organizationPolicy: VersionedOrganizationPolicy;
  /**
   *
   * @type {string}
   * @memberof OrganizationToWallet
   */
  policyName?: string | null;
  /**
   *
   * @type {VersionedTransactionPolicy}
   * @memberof OrganizationToWallet
   */
  transactionPolicy?: VersionedTransactionPolicy | null;
  /**
   *
   * @type {string}
   * @memberof OrganizationToWallet
   */
  walletId: string;
}
