// @phantom/openapi-wallet-service@0.1.11/model/partial-kms-user.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { PartialKmsAuthenticator } from "./partial-kms-authenticator";
// May contain unused imports in some cases
// @ts-ignore
import type { UserPolicy } from "./user-policy";

/**
 *
 * @export
 * @interface PartialKmsUser
 */
export interface PartialKmsUser {
  /**
   *
   * @type {Array<PartialKmsAuthenticator>}
   * @memberof PartialKmsUser
   */
  authenticators: Array<PartialKmsAuthenticator>;
  /**
   *
   * @type {number}
   * @memberof PartialKmsUser
   */
  expiresInMs?: number | null;
  /**
   *
   * @type {UserPolicy}
   * @memberof PartialKmsUser
   */
  policy: UserPolicy;
  /**
   *
   * @type {{ [key: string]: string; }}
   * @memberof PartialKmsUser
   */
  traits?: { [key: string]: string };
  /**
   *
   * @type {string}
   * @memberof PartialKmsUser
   */
  username: string;
}
