// @phantom/openapi-wallet-service@0.1.11/model/revoke-organization-access.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { RevokeOrganizationAccessRequest } from "./revoke-organization-access-request";

/**
 * Revoke access to a wallet from another organization.
 * @export
 * @interface RevokeOrganizationAccess
 */
export interface RevokeOrganizationAccess {
  /**
   *
   * @type {string}
   * @memberof RevokeOrganizationAccess
   */
  method: RevokeOrganizationAccessMethodEnum;
  /**
   * Revoke access to a wallet from another organization.
   * @type {RevokeOrganizationAccessRequest}
   * @memberof RevokeOrganizationAccess
   */
  params: RevokeOrganizationAccessRequest;
}

/**
 * @export
 * @enum {string}
 */
export enum RevokeOrganizationAccessMethodEnum {
  revokeOrganizationAccess = "revokeOrganizationAccess",
}
