// @phantom/openapi-wallet-service@0.1.11/model/sign-raw-payload-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { Algorithm } from "./algorithm";
// May contain unused imports in some cases
// @ts-ignore
import type { DerivationInfoSchema } from "./derivation-info-schema";

/**
 *
 * @export
 * @interface SignRawPayloadRequest
 */
export interface SignRawPayloadRequest {
  /**
   *
   * @type {Algorithm}
   * @memberof SignRawPayloadRequest
   */
  algorithm: Algorithm;
  /**
   *
   * @type {DerivationInfoSchema}
   * @memberof SignRawPayloadRequest
   */
  derivationInfo?: DerivationInfoSchema | null;
  /**
   *
   * @type {string}
   * @memberof SignRawPayloadRequest
   */
  organizationId: string;
  /**
   * Base64-encoded payload to sign
   * @type {string}
   * @memberof SignRawPayloadRequest
   */
  payload: string;
  /**
   *
   * @type {string}
   * @memberof SignRawPayloadRequest
   */
  policyName?: string | null;
  /**
   *
   * @type {string}
   * @memberof SignRawPayloadRequest
   */
  walletId: string;
}
