// @phantom/openapi-wallet-service@0.1.11/model/transfer-organization-wallet-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { VersionedOrganizationPolicy } from "./versioned-organization-policy";

/**
 *
 * @export
 * @interface TransferOrganizationWalletRequest
 */
export interface TransferOrganizationWalletRequest {
  /**
   *
   * @type {string}
   * @memberof TransferOrganizationWalletRequest
   */
  ownerOrganizationId: string;
  /**
   * If set, the owner organization policy will retain access to the wallet with this policy after the transfer.
   * @type {VersionedOrganizationPolicy}
   * @memberof TransferOrganizationWalletRequest
   */
  retainPolicy?: VersionedOrganizationPolicy | null;
  /**
   *
   * @type {string}
   * @memberof TransferOrganizationWalletRequest
   */
  targetOrganizationId: string;
  /**
   *
   * @type {string}
   * @memberof TransferOrganizationWalletRequest
   */
  walletId: string;
}
