// @phantom/openapi-wallet-service@0.1.11/model/transfer-organization-wallet.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { TransferOrganizationWalletRequest } from "./transfer-organization-wallet-request";

/**
 *
 * @export
 * @interface TransferOrganizationWallet
 */
export interface TransferOrganizationWallet {
  /**
   *
   * @type {string}
   * @memberof TransferOrganizationWallet
   */
  method: TransferOrganizationWalletMethodEnum;
  /**
   *
   * @type {TransferOrganizationWalletRequest}
   * @memberof TransferOrganizationWallet
   */
  params: TransferOrganizationWalletRequest;
}

/**
 * @export
 * @enum {string}
 */
export enum TransferOrganizationWalletMethodEnum {
  transferOrganizationWallet = "transferOrganizationWallet",
}
