// @phantom/openapi-wallet-service@0.1.11/model/update-authenticator.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { UpdateAuthenticatorRequest } from "./update-authenticator-request";

/**
 *
 * @export
 * @interface UpdateAuthenticator
 */
export interface UpdateAuthenticator {
  /**
   *
   * @type {string}
   * @memberof UpdateAuthenticator
   */
  method: UpdateAuthenticatorMethodEnum;
  /**
   *
   * @type {UpdateAuthenticatorRequest}
   * @memberof UpdateAuthenticator
   */
  params: UpdateAuthenticatorRequest;
}

/**
 * @export
 * @enum {string}
 */
export enum UpdateAuthenticatorMethodEnum {
  updateAuthenticator = "updateAuthenticator",
}
