// @phantom/openapi-wallet-service@0.1.11/model/update-user-in-organization-request.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

// May contain unused imports in some cases
// @ts-ignore
import type { UserPolicy } from "./user-policy";

/**
 * Request to update a user in an organization. Supports only changing user policy and traits
 * @export
 * @interface UpdateUserInOrganizationRequest
 */
export interface UpdateUserInOrganizationRequest {
  /**
   *
   * @type {{ [key: string]: string; }}
   * @memberof UpdateUserInOrganizationRequest
   */
  addTraits?: { [key: string]: string };
  /**
   *
   * @type {UserPolicy}
   * @memberof UpdateUserInOrganizationRequest
   */
  newPolicy?: UserPolicy | null;
  /**
   *
   * @type {string}
   * @memberof UpdateUserInOrganizationRequest
   */
  organizationId: string;
  /**
   *
   * @type {Array<string>}
   * @memberof UpdateUserInOrganizationRequest
   */
  removeTraits?: Array<string> | null;
  /**
   *
   * @type {string}
   * @memberof UpdateUserInOrganizationRequest
   */
  username: string;
}
