// @phantom/browser-injected-sdk@2.0.1/dist/auto-confirm/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/auto-confirm/index.ts
var auto_confirm_exports = {};
__export(auto_confirm_exports, {
  createAutoConfirmPlugin: () => createAutoConfirmPlugin
});
module.exports = __toCommonJS(auto_confirm_exports);

// src/errors.ts
var PHANTOM_NOT_DETECTED = "Phantom wallet not detected.";
var APP_PROVIDER_NOT_FOUND = "App provider not found on window.phantom.";

// src/extension/isInstalled.ts
function isInstalled() {
  try {
    const phantom = window?.phantom;
    return !!phantom;
  } catch (error) {
    return false;
  }
}

// src/auto-confirm/getProvider.ts
function getProvider() {
  if (!isInstalled()) {
    throw new Error(PHANTOM_NOT_DETECTED);
  }
  const provider = window.phantom.app;
  if (!provider) {
    throw new Error(APP_PROVIDER_NOT_FOUND);
  }
  return provider;
}

// src/auto-confirm/autoConfirmEnable.ts
var import_constants = require("@phantom/constants");
async function autoConfirmEnable(params) {
  const provider = getProvider();
  const transformedParams = params?.chains ? { chains: params.chains.map(import_constants.networkIdToInternalCaip) } : {};
  const result = await provider.request({
    method: "phantom_auto_confirm_enable",
    params: transformedParams
  });
  return {
    ...result,
    chains: result.chains.map(import_constants.internalCaipToNetworkId)
  };
}

// src/auto-confirm/autoConfirmDisable.ts
var import_constants2 = require("@phantom/constants");
async function autoConfirmDisable() {
  const provider = getProvider();
  const result = await provider.request({
    method: "phantom_auto_confirm_disable",
    params: {}
  });
  return {
    ...result,
    chains: result.chains.map(import_constants2.internalCaipToNetworkId)
  };
}

// src/auto-confirm/autoConfirmStatus.ts
var import_constants3 = require("@phantom/constants");
async function autoConfirmStatus() {
  const provider = getProvider();
  const result = await provider.request({
    method: "phantom_auto_confirm_status",
    params: {}
  });
  return {
    ...result,
    chains: result.chains.map(import_constants3.internalCaipToNetworkId)
  };
}

// src/auto-confirm/autoConfirmSupportedChains.ts
var import_constants4 = require("@phantom/constants");
async function autoConfirmSupportedChains() {
  const provider = getProvider();
  const result = await provider.request({
    method: "phantom_auto_confirm_supported_chains",
    params: {}
  });
  return {
    chains: result.chains.map(import_constants4.internalCaipToNetworkId)
  };
}

// src/auto-confirm/plugin.ts
var autoConfirm = {
  autoConfirmEnable,
  autoConfirmDisable,
  autoConfirmStatus,
  autoConfirmSupportedChains
};
function createAutoConfirmPlugin() {
  return {
    name: "autoConfirm",
    create: () => autoConfirm
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createAutoConfirmPlugin
});
