// @phantom/browser-injected-sdk@2.0.1/dist/auto-confirm/index.mjs
import {
  APP_PROVIDER_NOT_FOUND,
  PHANTOM_NOT_DETECTED,
  isInstalled
} from "../chunk-RGKAWEP4.mjs";

// src/auto-confirm/getProvider.ts
function getProvider() {
  if (!isInstalled()) {
    throw new Error(PHANTOM_NOT_DETECTED);
  }
  const provider = window.phantom.app;
  if (!provider) {
    throw new Error(APP_PROVIDER_NOT_FOUND);
  }
  return provider;
}

// src/auto-confirm/autoConfirmEnable.ts
import { networkIdToInternalCaip, internalCaipToNetworkId } from "@phantom/constants";
async function autoConfirmEnable(params) {
  const provider = getProvider();
  const transformedParams = params?.chains ? { chains: params.chains.map(networkIdToInternalCaip) } : {};
  const result = await provider.request({
    method: "phantom_auto_confirm_enable",
    params: transformedParams
  });
  return {
    ...result,
    chains: result.chains.map(internalCaipToNetworkId)
  };
}

// src/auto-confirm/autoConfirmDisable.ts
import { internalCaipToNetworkId as internalCaipToNetworkId2 } from "@phantom/constants";
async function autoConfirmDisable() {
  const provider = getProvider();
  const result = await provider.request({
    method: "phantom_auto_confirm_disable",
    params: {}
  });
  return {
    ...result,
    chains: result.chains.map(internalCaipToNetworkId2)
  };
}

// src/auto-confirm/autoConfirmStatus.ts
import { internalCaipToNetworkId as internalCaipToNetworkId3 } from "@phantom/constants";
async function autoConfirmStatus() {
  const provider = getProvider();
  const result = await provider.request({
    method: "phantom_auto_confirm_status",
    params: {}
  });
  return {
    ...result,
    chains: result.chains.map(internalCaipToNetworkId3)
  };
}

// src/auto-confirm/autoConfirmSupportedChains.ts
import { internalCaipToNetworkId as internalCaipToNetworkId4 } from "@phantom/constants";
async function autoConfirmSupportedChains() {
  const provider = getProvider();
  const result = await provider.request({
    method: "phantom_auto_confirm_supported_chains",
    params: {}
  });
  return {
    chains: result.chains.map(internalCaipToNetworkId4)
  };
}

// src/auto-confirm/plugin.ts
var autoConfirm = {
  autoConfirmEnable,
  autoConfirmDisable,
  autoConfirmStatus,
  autoConfirmSupportedChains
};
function createAutoConfirmPlugin() {
  return {
    name: "autoConfirm",
    create: () => autoConfirm
  };
}
export {
  createAutoConfirmPlugin
};
