// @phantom/browser-injected-sdk@2.0.1/dist/chunk-RGKAWEP4.mjs
var __accessCheck = (obj, member, msg) => {
  if (!member.has(obj))
    throw TypeError("Cannot " + msg);
};
var __privateAdd = (obj, member, value) => {
  if (member.has(obj))
    throw TypeError("Cannot add the same private member more than once");
  member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
};
var __privateMethod = (obj, member, method) => {
  __accessCheck(obj, member, "access private method");
  return method;
};

// src/extension/isInstalled.ts
function isInstalled() {
  try {
    const phantom = window?.phantom;
    return !!phantom;
  } catch (error) {
    return false;
  }
}

// src/errors.ts
var PHANTOM_NOT_DETECTED = "Phantom wallet not detected.";
var ETHEREUM_PROVIDER_NOT_FOUND = "Ethereum provider not found on window.phantom.";
var SOLANA_PROVIDER_NOT_FOUND = "Solana provider not found on window.phantom.";
var APP_PROVIDER_NOT_FOUND = "App provider not found on window.phantom.";

export {
  __privateAdd,
  __privateMethod,
  isInstalled,
  PHANTOM_NOT_DETECTED,
  ETHEREUM_PROVIDER_NOT_FOUND,
  SOLANA_PROVIDER_NOT_FOUND,
  APP_PROVIDER_NOT_FOUND
};
