// @phantom/parsers@2.0.1/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  deserializeSolanaTransaction: () => deserializeSolanaTransaction,
  parseSignMessageResponse: () => parseSignMessageResponse,
  parseSolanaKitTransactionToSolanaWeb3js: () => parseSolanaKitTransactionToSolanaWeb3js,
  parseSolanaSignedTransaction: () => parseSolanaSignedTransaction,
  parseToKmsTransaction: () => parseToKmsTransaction,
  parseTransactionResponse: () => parseTransactionResponse,
  validateEip712TypedData: () => validateEip712TypedData
});
module.exports = __toCommonJS(src_exports);
var import_base64url2 = require("@phantom/base64url");
var import_transactions = require("@solana/transactions");
var import_buffer2 = require("buffer");
var import_ethers = require("ethers");

// src/response-parsers.ts
var import_base64url = require("@phantom/base64url");
var import_constants = require("@phantom/constants");
var import_utils = require("@phantom/utils");
var import_web3 = require("@solana/web3.js");
var import_bs58 = __toESM(require("bs58"));
var import_buffer = require("buffer");
function parseSignMessageResponse(base64Response, networkId) {
  const networkPrefix = networkId.split(":")[0].toLowerCase();
  switch (networkPrefix) {
    case "solana":
      return parseSolanaSignatureResponse(base64Response);
    case "eip155":
    case "ethereum":
      return parseEVMSignatureResponse(base64Response);
    case "sui":
      return parseSuiSignatureResponse(base64Response);
    case "bip122":
    case "bitcoin":
      return parseBitcoinSignatureResponse(base64Response);
    default:
      return {
        signature: base64Response,
        rawSignature: base64Response
      };
  }
}
function parseTransactionResponse(base64RawTransaction, networkId, hash) {
  let rawTransaction = base64RawTransaction;
  if ((0, import_utils.isEthereumChain)(networkId)) {
    try {
      const txBytes = (0, import_base64url.base64urlDecode)(base64RawTransaction);
      rawTransaction = "0x" + import_buffer.Buffer.from(txBytes).toString("hex");
    } catch (error) {
      rawTransaction = base64RawTransaction.startsWith("0x") ? base64RawTransaction : "0x" + base64RawTransaction;
    }
  }
  if (hash) {
    return {
      hash,
      rawTransaction,
      blockExplorer: (0, import_constants.getExplorerUrl)(networkId, "transaction", hash)
    };
  } else {
    return {
      rawTransaction
    };
  }
}
function parseSolanaSignatureResponse(base64Response) {
  try {
    const signatureBytes = (0, import_base64url.base64urlDecode)(base64Response);
    const signature = import_bs58.default.encode(signatureBytes);
    return {
      signature,
      rawSignature: base64Response
    };
  } catch (error) {
    return {
      signature: base64Response,
      rawSignature: base64Response
    };
  }
}
function parseEVMSignatureResponse(base64Response) {
  const signatureBytes = (0, import_base64url.base64urlDecode)(base64Response);
  const signature = "0x" + import_buffer.Buffer.from(signatureBytes).toString("hex");
  return {
    signature,
    rawSignature: base64Response
  };
}
function parseSuiSignatureResponse(base64Response) {
  try {
    const signatureBytes = (0, import_base64url.base64urlDecode)(base64Response);
    const signature = import_buffer.Buffer.from(signatureBytes).toString("base64");
    return {
      signature,
      rawSignature: base64Response
    };
  } catch (error) {
    return {
      signature: base64Response,
      rawSignature: base64Response
    };
  }
}
function parseBitcoinSignatureResponse(base64Response) {
  try {
    const signatureBytes = (0, import_base64url.base64urlDecode)(base64Response);
    const signature = import_buffer.Buffer.from(signatureBytes).toString("hex");
    return {
      signature,
      rawSignature: base64Response
    };
  } catch (error) {
    return {
      signature: base64Response,
      rawSignature: base64Response
    };
  }
}
function parseSolanaSignedTransaction(base64RawTransaction) {
  try {
    const transactionBytes = (0, import_base64url.base64urlDecode)(base64RawTransaction);
    return deserializeSolanaTransaction(transactionBytes);
  } catch (error) {
    return null;
  }
}
function deserializeSolanaTransaction(transactionBytes) {
  try {
    const transaction = import_web3.Transaction.from(transactionBytes);
    return transaction;
  } catch (legacyError) {
    if (legacyError instanceof Error && legacyError.message.includes("Versioned messages")) {
      const versionedTransaction = import_web3.VersionedTransaction.deserialize(transactionBytes);
      return versionedTransaction;
    }
    throw legacyError;
  }
}

// src/index.ts
function validateEip712TypedData(data) {
  if (typeof data !== "object" || data === null) {
    throw new Error("typedData must be an object");
  }
  const obj = data;
  if (typeof obj.types !== "object" || obj.types === null || Array.isArray(obj.types)) {
    throw new Error("typedData.types must be an object mapping type names to field arrays");
  }
  if (typeof obj.primaryType !== "string" || !obj.primaryType) {
    throw new Error("typedData.primaryType must be a non-empty string");
  }
  if (typeof obj.domain !== "object" || obj.domain === null || Array.isArray(obj.domain)) {
    throw new Error("typedData.domain must be an object");
  }
  if (typeof obj.message !== "object" || obj.message === null || Array.isArray(obj.message)) {
    throw new Error("typedData.message must be an object");
  }
}
async function parseToKmsTransaction(transaction, networkId) {
  const networkPrefix = networkId.split(":")[0].toLowerCase();
  switch (networkPrefix) {
    case "solana":
      return parseSolanaTransactionToBase64Url(transaction);
    case "ethereum":
    case "eip155":
    case "polygon":
    case "optimism":
    case "arbitrum":
    case "base":
      return parseEVMTransactionToHex(transaction);
    case "sui":
      return await parseSuiTransactionToBase64Url(transaction);
    case "bitcoin":
      return parseBitcoinTransactionToBase64Url(transaction);
    default:
      throw new Error(`Unsupported network: ${networkPrefix}`);
  }
}
function parseSolanaTransactionToBase64Url(transaction) {
  if (transaction?.messageBytes != null) {
    return {
      parsed: (0, import_base64url2.base64urlEncode)(transaction.messageBytes),
      originalFormat: "@solana/kit"
    };
  }
  if (typeof transaction?.serialize === "function") {
    const serialized = transaction.serialize({
      requireAllSignatures: false,
      verifySignatures: false
    });
    return {
      parsed: (0, import_base64url2.base64urlEncode)(serialized),
      originalFormat: "@solana/web3.js"
    };
  }
  if (transaction instanceof Uint8Array) {
    return {
      parsed: (0, import_base64url2.base64urlEncode)(transaction),
      originalFormat: "bytes"
    };
  }
  if (typeof transaction === "string") {
    try {
      const bytes = import_buffer2.Buffer.from(transaction, "base64");
      return {
        parsed: (0, import_base64url2.base64urlEncode)(new Uint8Array(bytes)),
        originalFormat: "base64"
      };
    } catch {
      throw new Error("Unsupported Solana transaction format");
    }
  }
  throw new Error("Unsupported Solana transaction format");
}
function parseEVMTransactionToHex(transaction) {
  if (typeof transaction === "string" && transaction.startsWith("0x")) {
    return {
      parsed: transaction,
      originalFormat: "hex"
    };
  }
  if (transaction?.serialize && typeof transaction.serialize === "function") {
    const serialized = transaction.serialize();
    const hex = serialized.startsWith("0x") ? serialized : "0x" + serialized;
    return {
      parsed: hex,
      originalFormat: "ethers"
    };
  }
  if (transaction instanceof Uint8Array) {
    const hex = "0x" + import_buffer2.Buffer.from(transaction).toString("hex");
    return {
      parsed: hex,
      originalFormat: "bytes"
    };
  }
  if (transaction && typeof transaction === "object" && (transaction.to || transaction.data || transaction.from)) {
    try {
      const { from, gas, ...txForSerialization } = transaction;
      if (gas) {
        txForSerialization.gasLimit = gas;
      }
      if (!txForSerialization.gasLimit) {
        if (txForSerialization.to && txForSerialization.value && !txForSerialization.data) {
          txForSerialization.gasLimit = "0x5208";
        }
      }
      if (txForSerialization.to && typeof txForSerialization.to === "string") {
        try {
          txForSerialization.to = (0, import_ethers.getAddress)(txForSerialization.to);
        } catch {
          txForSerialization.to = txForSerialization.to.toLowerCase();
        }
      }
      const serialized = import_ethers.Transaction.from(txForSerialization).unsignedSerialized;
      const hex = serialized.startsWith("0x") ? serialized : "0x" + serialized;
      return {
        parsed: hex,
        originalFormat: "json"
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      const txKeys = transaction ? Object.keys(transaction).join(", ") : "N/A";
      const txValues = transaction ? JSON.stringify(transaction, null, 2) : "N/A";
      throw new Error(
        `Failed to RLP encode EVM transaction: ${errorMessage}.
Transaction keys: [${txKeys}].
Transaction: ${txValues}
Please ensure the transaction object includes required fields (to, value, chainId, gasLimit or gasPrice, etc.)`
      );
    }
  }
  throw new Error(
    "Unsupported EVM transaction format. Expected hex string, bytes, or transaction object with 'to', 'data', or 'from' fields."
  );
}
async function parseSuiTransactionToBase64Url(transaction) {
  if (transaction?.serialize && typeof transaction.serialize === "function") {
    const serialized = transaction.serialize();
    return {
      parsed: (0, import_base64url2.base64urlEncode)(serialized),
      originalFormat: "sui-sdk"
    };
  }
  if (transaction instanceof Uint8Array) {
    return {
      parsed: (0, import_base64url2.base64urlEncode)(transaction),
      originalFormat: "bytes"
    };
  }
  if (transaction?.build && typeof transaction.build === "function") {
    const built = await transaction.build();
    if (built?.serialize && typeof built.serialize === "function") {
      const serialized = built.serialize();
      return {
        parsed: (0, import_base64url2.base64urlEncode)(serialized),
        originalFormat: "transaction-block"
      };
    }
  }
  throw new Error("Unsupported Sui transaction format");
}
function parseBitcoinTransactionToBase64Url(transaction) {
  if (transaction?.toBuffer && typeof transaction.toBuffer === "function") {
    const buffer = transaction.toBuffer();
    return {
      parsed: (0, import_base64url2.base64urlEncode)(new Uint8Array(buffer)),
      originalFormat: "bitcoinjs-lib"
    };
  }
  if (transaction instanceof Uint8Array) {
    return {
      parsed: (0, import_base64url2.base64urlEncode)(transaction),
      originalFormat: "bytes"
    };
  }
  if (typeof transaction === "string") {
    const bytes = new Uint8Array(import_buffer2.Buffer.from(transaction, "hex"));
    return {
      parsed: (0, import_base64url2.base64urlEncode)(bytes),
      originalFormat: "hex"
    };
  }
  throw new Error("Unsupported Bitcoin transaction format");
}
function parseSolanaKitTransactionToSolanaWeb3js(transaction) {
  const serialized = (0, import_transactions.getTransactionEncoder)().encode(transaction);
  const fakeVersioned = {
    serialize() {
      return new Uint8Array(serialized);
    }
  };
  return fakeVersioned;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  deserializeSolanaTransaction,
  parseSignMessageResponse,
  parseSolanaKitTransactionToSolanaWeb3js,
  parseSolanaSignedTransaction,
  parseToKmsTransaction,
  parseTransactionResponse,
  validateEip712TypedData
});
