// @phantom/api-key-stamper@2.0.1/dist/index.mjs
// src/index.ts
import { base64urlEncode } from "@phantom/base64url";
import { createKeyPairFromSecret, signWithSecret } from "@phantom/crypto";
import bs58 from "bs58";
import { DEFAULT_AUTHENTICATOR_ALGORITHM } from "@phantom/constants";
var ApiKeyStamper = class {
  constructor(config) {
    this.algorithm = DEFAULT_AUTHENTICATOR_ALGORITHM;
    // Use the same algorithm as the keypair
    this.type = "PKI";
    this.keypair = createKeyPairFromSecret(config.apiSecretKey);
  }
  /**
   * Create X-Phantom-Stamp header value
   * @param params - Parameters object with data to sign
   * @returns Complete X-Phantom-Stamp header value
   */
  async stamp(params) {
    const { data } = params;
    const signature = signWithSecret(this.keypair.secretKey, data);
    const signatureBase64url = base64urlEncode(signature);
    const stampType = params.type || this.type;
    const stampData = stampType === "PKI" ? {
      publicKey: base64urlEncode(bs58.decode(this.keypair.publicKey)),
      signature: signatureBase64url,
      kind: "PKI",
      algorithm: this.algorithm
    } : {
      kind: "OIDC",
      idToken: params.type === "OIDC" ? params.idToken : this.idToken,
      publicKey: base64urlEncode(bs58.decode(this.keypair.publicKey)),
      salt: params.type === "OIDC" ? params.salt : this.salt,
      algorithm: this.algorithm,
      signature: signatureBase64url
    };
    const stampJson = JSON.stringify(stampData);
    return Promise.resolve(base64urlEncode(new TextEncoder().encode(stampJson)));
  }
};
export {
  ApiKeyStamper
};
