// @phantom/api-key-stamper@2.0.1/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ApiKeyStamper: () => ApiKeyStamper
});
module.exports = __toCommonJS(src_exports);
var import_base64url = require("@phantom/base64url");
var import_crypto = require("@phantom/crypto");
var import_bs58 = __toESM(require("bs58"));
var import_constants = require("@phantom/constants");
var ApiKeyStamper = class {
  constructor(config) {
    this.algorithm = import_constants.DEFAULT_AUTHENTICATOR_ALGORITHM;
    // Use the same algorithm as the keypair
    this.type = "PKI";
    this.keypair = (0, import_crypto.createKeyPairFromSecret)(config.apiSecretKey);
  }
  /**
   * Create X-Phantom-Stamp header value
   * @param params - Parameters object with data to sign
   * @returns Complete X-Phantom-Stamp header value
   */
  async stamp(params) {
    const { data } = params;
    const signature = (0, import_crypto.signWithSecret)(this.keypair.secretKey, data);
    const signatureBase64url = (0, import_base64url.base64urlEncode)(signature);
    const stampType = params.type || this.type;
    const stampData = stampType === "PKI" ? {
      publicKey: (0, import_base64url.base64urlEncode)(import_bs58.default.decode(this.keypair.publicKey)),
      signature: signatureBase64url,
      kind: "PKI",
      algorithm: this.algorithm
    } : {
      kind: "OIDC",
      idToken: params.type === "OIDC" ? params.idToken : this.idToken,
      publicKey: (0, import_base64url.base64urlEncode)(import_bs58.default.decode(this.keypair.publicKey)),
      salt: params.type === "OIDC" ? params.salt : this.salt,
      algorithm: this.algorithm,
      signature: signatureBase64url
    };
    const stampJson = JSON.stringify(stampData);
    return Promise.resolve((0, import_base64url.base64urlEncode)(new TextEncoder().encode(stampJson)));
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApiKeyStamper
});
