// @phantom/auth2@2.0.1/dist/index.mjs
// src/Auth2KmsRpcClient.ts
import {
  Configuration,
  KMSRPCApi,
  GetOrCreatePhantomOrganizationMethodEnum,
  GetOrCreateWalletWithTagMethodEnum
} from "@phantom/openapi-wallet-service";
import axios from "axios";
import { Buffer } from "buffer";
var DEFAULT_KMS_API_VERSION = "2025-11-24";
var Auth2KmsRpcClient = class {
  constructor(stamper, options) {
    this.stamper = stamper;
    const axiosInstance = axios.create();
    axiosInstance.interceptors.request.use(async (config) => {
      await this.stamper.maybeRefreshTokens();
      config.headers = config.headers || {};
      config.headers["x-app-id"] = options.appId;
      config.headers["x-api-version"] = DEFAULT_KMS_API_VERSION;
      config.headers["authorization"] = this.stamper.bearerToken;
      const authUserId = this.stamper.auth2Token?.sub;
      if (authUserId) {
        config.headers["x-auth-user-id"] = authUserId;
      }
      const requestBody = typeof config.data === "string" ? config.data : config.data === void 0 ? "" : JSON.stringify(config.data);
      const stamp = await this.stamper.stamp({ data: Buffer.from(requestBody, "utf-8") });
      config.headers["x-phantom-stamp"] = stamp;
      return config;
    });
    const configuration = new Configuration({ basePath: options.apiBaseUrl });
    this.kmsApi = new KMSRPCApi(configuration, options.apiBaseUrl, axiosInstance);
  }
  async postKmsRpc(request) {
    const response = await this.kmsApi.postKmsRpc(request);
    const rpcBody = response.data;
    if ("error" in rpcBody) {
      throw new Error(`KMS RPC error: ${JSON.stringify(rpcBody.error)}`);
    }
    return rpcBody.result;
  }
  async getOrCreatePhantomOrganization(publicKey) {
    return await this.postKmsRpc({
      method: GetOrCreatePhantomOrganizationMethodEnum.getOrCreatePhantomOrganization,
      params: { publicKey },
      timestampMs: Date.now()
    });
  }
  // listPendingMigrations is not yet in @phantom/openapi-wallet-service
  async listPendingMigrations(organizationId) {
    return await this.postKmsRpc({
      method: "listPendingMigrations",
      params: { organizationId },
      timestampMs: Date.now()
    });
  }
  // completeWalletTransfer is not yet in @phantom/openapi-wallet-service
  async completeWalletTransfer(args) {
    return await this.postKmsRpc({
      method: "completeWalletTransfer",
      params: args,
      timestampMs: Date.now()
    });
  }
  async getOrCreateWalletWithTag(args) {
    return await this.postKmsRpc({
      method: GetOrCreateWalletWithTagMethodEnum.getOrCreateWalletWithTag,
      params: args,
      timestampMs: Date.now()
    });
  }
};

// src/Auth2Stamper.ts
import bs58 from "bs58";
import { base64urlEncode } from "@phantom/base64url";
import { Algorithm } from "@phantom/sdk-types";

// src/Auth2Token.ts
import { jwtDecode } from "jwt-decode";
var Auth2Token = class {
  constructor(accessToken) {
    this.WALLET_URN_PREFIX = "urn:phantom:wallet:";
    this.WALLET_TAG_URN_PREFIX = "urn:phantom:wallet-tag:";
    this._claims = decodeJwtClaims(accessToken);
    this._a2tClaims = decodeJwtClaims(this._claims.ext.a2t);
    const walletAud = this._claims.aud.find((aud) => aud.startsWith(this.WALLET_URN_PREFIX));
    if (walletAud) {
      const [id, derivationIndex] = walletAud.replace(this.WALLET_URN_PREFIX, "").split(":");
      this._identity = {
        id,
        derivationIndex: Number(derivationIndex)
      };
    }
    const walletTagAud = this._claims.aud.find((aud) => aud.startsWith(this.WALLET_TAG_URN_PREFIX));
    if (walletTagAud) {
      this._walletTag = walletTagAud.replace(this.WALLET_TAG_URN_PREFIX, "");
    }
  }
  static fromAccessToken(accessToken) {
    return new this(accessToken);
  }
  get sub() {
    return this._claims.sub;
  }
  get clientId() {
    return this._claims.client_id;
  }
  get wallet() {
    return this._identity;
  }
  get walletTag() {
    return this._walletTag;
  }
  get a2t() {
    if (this._a2tClaims.exp < Date.now() / 1e3) {
      throw new Auth2TokenExpiredError();
    }
    return this._claims.ext.a2t;
  }
};
function decodeJwtClaims(jwt) {
  return jwtDecode(jwt);
}
var Auth2TokenExpiredError = class extends Error {
  constructor() {
    super("Auth2 token expired");
    this.name = "Auth2TokenExpiredError";
  }
};

// src/tokenExchange.ts
async function exchangeAuthCode(options) {
  return _postTokenRequest(
    options.authApiBaseUrl,
    new URLSearchParams({
      grant_type: "authorization_code",
      client_id: options.clientId,
      redirect_uri: options.redirectUri,
      code: options.code,
      code_verifier: options.codeVerifier
    })
  );
}
async function refreshToken(options) {
  const params = new URLSearchParams({
    grant_type: "refresh_token",
    client_id: options.clientId,
    refresh_token: options.refreshToken
  });
  if (options.redirectUri.trim().length > 0) {
    params.set("redirect_uri", options.redirectUri);
  }
  const result = await _postTokenRequest(options.authApiBaseUrl, params);
  if (!result.refreshToken) {
    throw new Error("Auth2 refresh token request did not return a refresh_token.");
  }
  return { ...result, refreshToken: result.refreshToken };
}
async function _postTokenRequest(authApiBaseUrl, body) {
  const tokenEndpoint = new URL("/oauth2/token", authApiBaseUrl).toString();
  const tokenResponse = await fetch(tokenEndpoint, {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: body.toString()
  });
  if (!tokenResponse.ok) {
    const text = await tokenResponse.text();
    throw new Error(`Auth2 token request failed (${tokenResponse.status} ${tokenResponse.statusText}): ${text}`);
  }
  const tokens = await tokenResponse.json();
  const accessToken = tokens.access_token;
  if (!accessToken) {
    throw new Error("Auth2 token request did not return an access_token.");
  }
  return {
    accessToken,
    idType: tokens.token_type ?? "Bearer",
    expiresInMs: tokens.expires_in != null ? tokens.expires_in * 1e3 : 0,
    refreshToken: tokens.refresh_token
  };
}

// src/Auth2Stamper.ts
var TOKEN_REFRESH_BUFFER_MS = 15 * 60 * 1e3;
var Auth2Stamper = class {
  constructor(storage, refreshConfig) {
    this.storage = storage;
    this.refreshConfig = refreshConfig;
    this._keyPair = null;
    this._keyInfo = null;
    this._idType = null;
    this._accessToken = null;
    this._refreshToken = null;
    this._tokenExpiresAt = null;
    this.refreshingTokensPromise = null;
    this.algorithm = Algorithm.secp256r1;
    this.type = "OIDC";
  }
  get bearerToken() {
    if (!this._idType || !this._accessToken) {
      return null;
    }
    return `${this._idType} ${this._accessToken}`;
  }
  get auth2Token() {
    if (!this._accessToken) {
      return null;
    }
    return Auth2Token.fromAccessToken(this._accessToken);
  }
  async init() {
    if (this.storage.open) {
      await this.storage.open();
    }
    const stored = await this.storage.load();
    if (stored) {
      this._keyPair = stored.keyPair;
      this._keyInfo = stored.keyInfo;
      if (stored.idType) {
        this._idType = stored.idType;
      }
      if (stored.accessToken) {
        this._accessToken = stored.accessToken;
      }
      if (stored.refreshToken) {
        this._refreshToken = stored.refreshToken;
      }
      if (stored.tokenExpiresAt) {
        this._tokenExpiresAt = stored.tokenExpiresAt;
      }
      return this._keyInfo;
    }
    return this.generateAndStore();
  }
  getKeyInfo() {
    return this._keyInfo;
  }
  getCryptoKeyPair() {
    return this._keyPair;
  }
  /**
   * Arms the stamper with the access token data for subsequent KMS stamp() calls.
   *
   * Persists the tokens alongside the key pair so that auto-connect can
   * restore them on the next launch without a new login.
   */
  async setTokens({
    accessToken,
    idType,
    refreshToken: refreshToken2,
    expiresInMs
  }) {
    this._idType = idType;
    this._accessToken = accessToken;
    this._refreshToken = refreshToken2 ?? null;
    this._tokenExpiresAt = expiresInMs != null ? Date.now() + expiresInMs : null;
    const existing = await this.storage.load();
    if (existing) {
      await this.storage.save({
        ...existing,
        accessToken,
        idType,
        refreshToken: refreshToken2,
        tokenExpiresAt: this._tokenExpiresAt ?? void 0
      });
    }
  }
  /**
   * Checks if tokens should be refreshed and performs a refresh if needed.
   * Returns true if a refresh succeeded, false otherwise.
   */
  async maybeRefreshTokens() {
    this.refreshConfig?.logger?.debug?.("Auth2Stamper.maybeRefreshTokens() called");
    if (this.refreshingTokensPromise) {
      this.refreshConfig?.logger?.debug?.("Auth2Stamper.maybeRefreshTokens(): awaiting in-flight refresh");
      return this.refreshingTokensPromise;
    }
    if (!this.refreshConfig) {
      return false;
    }
    if (!this._refreshToken) {
      this.refreshConfig.logger?.debug?.("Auth2Stamper.maybeRefreshTokens(): skipped \u2014 missing refresh token");
      return false;
    }
    if (!this._tokenExpiresAt) {
      this.refreshConfig.logger?.debug?.("Auth2Stamper.maybeRefreshTokens(): skipped \u2014 missing token expiry");
      return false;
    }
    if (Date.now() < this._tokenExpiresAt - TOKEN_REFRESH_BUFFER_MS) {
      this.refreshConfig.logger?.debug?.("Auth2Stamper.maybeRefreshTokens(): skipped \u2014 token not near expiry");
      return false;
    }
    const refreshConfig = this.refreshConfig;
    const refreshToken2 = this._refreshToken;
    this.refreshingTokensPromise = (async () => {
      try {
        refreshConfig.logger?.info?.("Auth2Stamper.maybeRefreshTokens(): attempting token refresh");
        const refreshed = await refreshToken({
          authApiBaseUrl: refreshConfig.authApiBaseUrl,
          clientId: refreshConfig.clientId,
          redirectUri: refreshConfig.redirectUri,
          refreshToken: refreshToken2
        });
        await this.setTokens({
          accessToken: refreshed.accessToken,
          idType: refreshed.idType,
          refreshToken: refreshed.refreshToken,
          expiresInMs: refreshed.expiresInMs
        });
        refreshConfig.logger?.info?.("Auth2Stamper.maybeRefreshTokens(): token refresh succeeded");
        return true;
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        refreshConfig.logger?.error?.(`Auth2Stamper.maybeRefreshTokens(): token refresh failed: ${message}`);
        return false;
      } finally {
        this.refreshingTokensPromise = null;
      }
    })();
    return this.refreshingTokensPromise;
  }
  async stamp(params) {
    this.refreshConfig?.logger?.debug?.("Auth2Stamper.stamp(): called");
    if (!this._keyPair || !this._keyInfo) {
      throw new Error("Auth2Stamper not initialized. Call init() first.");
    }
    this.refreshConfig?.logger?.debug?.("Auth2Stamper.stamp(): awaiting maybeRefreshTokens()");
    await this.maybeRefreshTokens();
    this.refreshConfig?.logger?.debug?.("Auth2Stamper.stamp(): maybeRefreshTokens() completed");
    if (!this.auth2Token) {
      throw new Error("Auth2Stamper not initialized. Call init() first.");
    }
    const signatureRaw = await crypto.subtle.sign(
      { name: "ECDSA", hash: "SHA-256" },
      this._keyPair.privateKey,
      new Uint8Array(params.data)
    );
    const rawPublicKey = bs58.decode(this._keyInfo.publicKey);
    const stampData = {
      kind: this.type,
      idToken: this.auth2Token.a2t,
      publicKey: base64urlEncode(rawPublicKey),
      algorithm: this.algorithm,
      // The P-256 ephemeral key is unique per wallet, so no additional salt is needed.
      salt: "",
      signature: base64urlEncode(new Uint8Array(signatureRaw))
    };
    return base64urlEncode(new TextEncoder().encode(JSON.stringify(stampData)));
  }
  async resetKeyPair() {
    await this.clear();
    return this.generateAndStore();
  }
  async clear() {
    await this.storage.clear();
    this._keyPair = null;
    this._keyInfo = null;
    this._accessToken = null;
    this._idType = null;
    this._refreshToken = null;
    this._tokenExpiresAt = null;
  }
  // eslint-disable-next-line @typescript-eslint/require-await
  async rotateKeyPair() {
    throw new Error("rotateKeyPair is not supported for Auth2Stamper");
  }
  // eslint-disable-next-line @typescript-eslint/require-await
  async commitRotation(_authenticatorId) {
    throw new Error("commitRotation is not supported for Auth2Stamper");
  }
  // eslint-disable-next-line @typescript-eslint/require-await
  async rollbackRotation() {
    throw new Error("rollbackRotation is not supported for Auth2Stamper");
  }
  async generateAndStore() {
    const keyPair = await crypto.subtle.generateKey(
      { name: "ECDSA", namedCurve: "P-256" },
      this.storage.requiresExtractableKeys,
      ["sign", "verify"]
    );
    const rawPublicKey = new Uint8Array(await crypto.subtle.exportKey("raw", keyPair.publicKey));
    const publicKeyBase58 = bs58.encode(rawPublicKey);
    const keyIdBuffer = await crypto.subtle.digest("SHA-256", rawPublicKey.buffer);
    const keyId = base64urlEncode(new Uint8Array(keyIdBuffer)).substring(0, 16);
    this._keyPair = keyPair;
    this._keyInfo = { keyId, publicKey: publicKeyBase58, createdAt: Date.now() };
    await this.storage.save({ keyPair, keyInfo: this._keyInfo });
    return this._keyInfo;
  }
};

// src/auth2Flow.ts
import { base64urlEncode as base64urlEncode3 } from "@phantom/base64url";
import { sha256 } from "@phantom/crypto";
import {
  DerivationInfoCurveEnum,
  DerivationInfoAddressFormatEnum
} from "@phantom/openapi-wallet-service";
import bs582 from "bs58";

// src/jar.ts
import { base64urlDecode, base64urlEncode as base64urlEncode2 } from "@phantom/base64url";
var AUTH2_JAR_ALGORITHM = "ES256";
var AUTH2_JAR_TYP = "oauth-authz-req+jwt";
var textEncoder = new TextEncoder();
async function createAuth2RequestJar(input) {
  const publicJwk = await globalThis.crypto.subtle.exportKey("jwk", input.keyPair.publicKey);
  if (!_isValidPublicJwk(publicJwk)) {
    throw new Error("Unable to export a valid P-256 public JWK for JAR header.");
  }
  const header = {
    alg: AUTH2_JAR_ALGORITHM,
    typ: AUTH2_JAR_TYP,
    jwk: {
      kty: "EC",
      crv: "P-256",
      x: publicJwk.x,
      y: publicJwk.y
    }
  };
  const encodedHeader = base64urlEncode2(textEncoder.encode(JSON.stringify(header)));
  const encodedPayload = base64urlEncode2(textEncoder.encode(JSON.stringify(input.payload)));
  const signingInput = `${encodedHeader}.${encodedPayload}`;
  const signature = await globalThis.crypto.subtle.sign(
    { name: "ECDSA", hash: "SHA-256" },
    input.keyPair.privateKey,
    textEncoder.encode(signingInput)
  );
  return `${signingInput}.${base64urlEncode2(new Uint8Array(signature))}`;
}
function _isValidPublicJwk(jwk) {
  return jwk.kty === "EC" && jwk.crv === "P-256" && typeof jwk.x === "string" && typeof jwk.y === "string";
}

// src/auth2Flow.ts
var DEFAULT_SCOPE = "openid offline_access";
var DAPP_WALLET_DERIVATIONS = [
  {
    curve: DerivationInfoCurveEnum.ed25519,
    derivationPath: "m/44'/501'/0'/0'",
    addressFormat: DerivationInfoAddressFormatEnum.solana
  },
  {
    curve: DerivationInfoCurveEnum.secp256k1,
    derivationPath: "m/44'/60'/0'/0/0",
    addressFormat: DerivationInfoAddressFormatEnum.ethereum
  },
  {
    curve: DerivationInfoCurveEnum.secp256k1,
    derivationPath: "m/84'/0'/0'/0",
    addressFormat: DerivationInfoAddressFormatEnum.bitcoinSegwit
  },
  {
    curve: DerivationInfoCurveEnum.ed25519,
    derivationPath: "m/44'/784'/0'/0'/0'",
    addressFormat: DerivationInfoAddressFormatEnum.sui
  }
];
var DAPP_WALLET_MNEMONIC_LENGTH = 24;
var DAPP_WALLET_DERIVATION_INDEX = 0;
async function prepareAuth2Flow(options) {
  const { stamper, auth2Options } = options;
  if (!stamper.getKeyInfo()) {
    await stamper.init();
  }
  const keyPair = stamper.getCryptoKeyPair();
  if (!keyPair) {
    throw new Error("Stamper key pair not found.");
  }
  const codeVerifier = createCodeVerifier();
  const url = await createConnectStartUrl({
    keyPair,
    connectLoginUrl: auth2Options.connectLoginUrl,
    clientId: auth2Options.clientId,
    redirectUri: auth2Options.redirectUri,
    sessionId: options.sessionId,
    provider: options.provider,
    codeVerifier,
    salt: ""
  });
  return { url, codeVerifier, keyPair };
}
function createCodeVerifier() {
  const bytes = new Uint8Array(64);
  crypto.getRandomValues(bytes);
  return base64urlEncode3(bytes).slice(0, 96);
}
async function createConnectStartUrl({
  keyPair,
  connectLoginUrl,
  clientId,
  redirectUri,
  sessionId,
  provider,
  codeVerifier,
  salt
}) {
  const nonce = await _deriveNonce(keyPair, salt);
  const codeChallenge = await _createCodeChallenge(codeVerifier);
  const nowSeconds = Math.floor(Date.now() / 1e3);
  const jarPayload = {
    aud: connectLoginUrl,
    iat: nowSeconds,
    exp: nowSeconds + 5 * 60,
    // 5 minutes
    client_id: clientId,
    redirect_uri: redirectUri,
    scope: DEFAULT_SCOPE,
    nonce,
    code_challenge: codeChallenge,
    code_challenge_method: "S256",
    ...provider && provider !== "phantom" && provider !== "device" && {
      login_hint: `${provider}:auth2`
    },
    // Use session_id as the OAuth state so it comes back in the callback URL
    // and can be validated without an extra sessionStorage entry.
    state: sessionId,
    should_migrate: true
  };
  const jar = await createAuth2RequestJar({
    payload: jarPayload,
    keyPair
  });
  const url = new URL(connectLoginUrl);
  url.hash = `jar=${jar}`;
  return url.toString();
}
async function _deriveNonce(keyPair, salt) {
  const publicKey = await crypto.subtle.exportKey("raw", keyPair.publicKey);
  const rawPublicKey = new Uint8Array(publicKey);
  const saltBytes = new TextEncoder().encode(salt);
  const combined = new Uint8Array(rawPublicKey.length + saltBytes.length);
  combined.set(rawPublicKey);
  combined.set(saltBytes, rawPublicKey.length);
  return base64urlEncode3(await sha256(combined));
}
async function _createCodeChallenge(codeVerifier) {
  return base64urlEncode3(await sha256(new TextEncoder().encode(codeVerifier)));
}
function validateAuth2Callback(options) {
  const { getParam, expectedSessionId } = options;
  const state = getParam("state");
  if (state !== expectedSessionId) {
    throw new Error("Auth2 state mismatch \u2014 possible CSRF attack.");
  }
  const error = getParam("error");
  if (error) {
    const description = getParam("error_description");
    throw new Error(`Auth2 callback error: ${description ?? error}`);
  }
  const code = getParam("code");
  if (!code) {
    throw new Error("Auth2 callback missing authorization code");
  }
  return code;
}
async function completeAuth2Exchange(options) {
  const { stamper, kms, auth2Options } = options;
  const { accessToken, idType, expiresInMs, refreshToken: refreshToken2 } = await exchangeAuthCode({
    authApiBaseUrl: auth2Options.authApiBaseUrl,
    clientId: auth2Options.clientId,
    redirectUri: auth2Options.redirectUri,
    code: options.code,
    codeVerifier: options.codeVerifier
  });
  await stamper.setTokens({
    accessToken,
    idType,
    refreshToken: refreshToken2,
    expiresInMs
  });
  const auth2Token = Auth2Token.fromAccessToken(accessToken);
  const { organizationId, wallet } = await _getOrMigrateWallet({ stamper, kms, auth2Token });
  return {
    walletId: wallet.id,
    organizationId,
    provider: options.provider,
    accountDerivationIndex: wallet.derivationIndex,
    expiresInMs,
    authUserId: auth2Token.sub,
    bearerToken: `${idType} ${accessToken}`
  };
}
async function _getOrMigrateWallet({
  stamper,
  kms,
  auth2Token
}) {
  const keyInfo = stamper.getKeyInfo();
  if (!keyInfo) {
    throw new Error("Stamper key pair not found.");
  }
  const publicKey = base64urlEncode3(bs582.decode(keyInfo.publicKey));
  const { organizationId } = await kms.getOrCreatePhantomOrganization(publicKey);
  if (auth2Token.wallet) {
    const pendingMigrations = await kms.listPendingMigrations(organizationId);
    if (pendingMigrations.pendingMigrations) {
      for (const migration of pendingMigrations.pendingMigrations) {
        if (migration.migrationId) {
          try {
            await kms.completeWalletTransfer({ organizationId, migrationId: migration.migrationId });
          } catch (error) {
            console.error(`Failed to complete wallet transfer for migration ${migration.migrationId}:`, error);
          }
        }
      }
    }
    return {
      organizationId,
      wallet: auth2Token.wallet
    };
  }
  const appWallet = await _getOrCreateAppWallet({
    kms,
    organizationId,
    clientId: auth2Token.clientId
  });
  return {
    organizationId,
    wallet: {
      id: appWallet.walletId,
      derivationIndex: DAPP_WALLET_DERIVATION_INDEX
    }
  };
}
async function _getOrCreateAppWallet({
  kms,
  organizationId,
  clientId
}) {
  return await kms.getOrCreateWalletWithTag({
    organizationId,
    walletName: "App Wallet",
    tag: clientId,
    accounts: DAPP_WALLET_DERIVATIONS,
    mnemonicLength: DAPP_WALLET_MNEMONIC_LENGTH
  });
}
export {
  Auth2KmsRpcClient,
  Auth2Stamper,
  Auth2Token,
  Auth2TokenExpiredError,
  _createCodeChallenge,
  _deriveNonce,
  _getOrCreateAppWallet,
  completeAuth2Exchange,
  createCodeVerifier,
  createConnectStartUrl,
  decodeJwtClaims,
  exchangeAuthCode,
  prepareAuth2Flow,
  refreshToken,
  validateAuth2Callback
};
