// @phantom/crypto@2.0.1/dist/index.mjs
// src/index.ts
import nacl from "tweetnacl";
import bs58 from "bs58";
import { Buffer } from "buffer";
async function sha256(input) {
  const digest = await crypto.subtle.digest("SHA-256", input);
  return new Uint8Array(digest);
}
function generateKeyPair() {
  const keypair = nacl.sign.keyPair();
  return {
    publicKey: bs58.encode(keypair.publicKey),
    secretKey: bs58.encode(keypair.secretKey)
  };
}
function createKeyPairFromSecret(b58PrivateKey) {
  const secretKeyBytes = bs58.decode(b58PrivateKey);
  const keypair = nacl.sign.keyPair.fromSecretKey(secretKeyBytes);
  return {
    publicKey: bs58.encode(keypair.publicKey),
    secretKey: bs58.encode(keypair.secretKey)
  };
}
function signWithSecret(secretKey, data) {
  const secretKeyBytes = typeof secretKey === "string" ? bs58.decode(secretKey) : secretKey;
  let dataBytes;
  if (typeof data === "string") {
    dataBytes = new TextEncoder().encode(data);
  } else if (data instanceof Buffer) {
    dataBytes = new Uint8Array(data);
  } else {
    dataBytes = data;
  }
  return nacl.sign.detached(dataBytes, secretKeyBytes);
}
export {
  createKeyPairFromSecret,
  generateKeyPair,
  sha256,
  signWithSecret
};
