// @phantom/crypto@2.0.1/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  createKeyPairFromSecret: () => createKeyPairFromSecret,
  generateKeyPair: () => generateKeyPair,
  sha256: () => sha256,
  signWithSecret: () => signWithSecret
});
module.exports = __toCommonJS(src_exports);
var import_tweetnacl = __toESM(require("tweetnacl"));
var import_bs58 = __toESM(require("bs58"));
var import_buffer = require("buffer");
async function sha256(input) {
  const digest = await crypto.subtle.digest("SHA-256", input);
  return new Uint8Array(digest);
}
function generateKeyPair() {
  const keypair = import_tweetnacl.default.sign.keyPair();
  return {
    publicKey: import_bs58.default.encode(keypair.publicKey),
    secretKey: import_bs58.default.encode(keypair.secretKey)
  };
}
function createKeyPairFromSecret(b58PrivateKey) {
  const secretKeyBytes = import_bs58.default.decode(b58PrivateKey);
  const keypair = import_tweetnacl.default.sign.keyPair.fromSecretKey(secretKeyBytes);
  return {
    publicKey: import_bs58.default.encode(keypair.publicKey),
    secretKey: import_bs58.default.encode(keypair.secretKey)
  };
}
function signWithSecret(secretKey, data) {
  const secretKeyBytes = typeof secretKey === "string" ? import_bs58.default.decode(secretKey) : secretKey;
  let dataBytes;
  if (typeof data === "string") {
    dataBytes = new TextEncoder().encode(data);
  } else if (data instanceof import_buffer.Buffer) {
    dataBytes = new Uint8Array(data);
  } else {
    dataBytes = data;
  }
  return import_tweetnacl.default.sign.detached(dataBytes, secretKeyBytes);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createKeyPairFromSecret,
  generateKeyPair,
  sha256,
  signWithSecret
});
