// @phantom/indexed-db-stamper@2.0.1/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  IndexedDbStamper: () => IndexedDbStamper
});
module.exports = __toCommonJS(src_exports);
var import_base64url = require("@phantom/base64url");
var import_constants = require("@phantom/constants");
var import_bs58 = __toESM(require("bs58"));
var import_sdk_types = require("@phantom/sdk-types");
var WEB_CRYPTO_ALGORITHM_CONFIGS = {
  [import_constants.DEFAULT_AUTHENTICATOR_ALGORITHM]: {
    generateParams: { name: "Ed25519" },
    signParams: { name: "Ed25519" }
  },
  [import_sdk_types.Algorithm.secp256r1]: {
    generateParams: { name: "ECDSA", namedCurve: "P-256" },
    signParams: { name: "ECDSA", hash: "SHA-256" }
  }
};
var IndexedDbStamper = class {
  // Optional for PKI, required for OIDC
  constructor(config = {}) {
    this.db = null;
    this.activeKeyPairRecord = null;
    this.pendingKeyPairRecord = null;
    this.algorithm = import_constants.DEFAULT_AUTHENTICATOR_ALGORITHM;
    // The type of stamper, can be changed at any time
    this.type = "PKI";
    if (typeof window === "undefined" || !window.indexedDB) {
      throw new Error("IndexedDbStamper requires a browser environment with IndexedDB support");
    }
    this.dbName = config.dbName || "phantom-indexed-db-stamper";
    this.storeName = config.storeName || "crypto-keys";
    this.keyName = config.keyName || "signing-key";
    this.type = config.type || "PKI";
    this.idToken = config.idToken;
    this.salt = config.salt;
  }
  /**
   * Initialize the stamper by opening IndexedDB and retrieving or generating keys.
   * Automatically selects the best supported algorithm.
   */
  async init() {
    await this.openDB();
    this.activeKeyPairRecord = await this.loadActiveKeyPairRecord();
    if (this.activeKeyPairRecord) {
      this.algorithm = this.activeKeyPairRecord.keyPair.privateKey.algorithm.name === "Ed25519" ? import_sdk_types.Algorithm.ed25519 : import_sdk_types.Algorithm.secp256r1;
    } else {
      this.algorithm = await this.getSupportedAlgorithm();
      this.activeKeyPairRecord = await this.generateAndStoreNewKeyPair("active");
    }
    this.pendingKeyPairRecord = await this.loadPendingKeyPairRecord();
    return this.activeKeyPairRecord.keyInfo;
  }
  async getSupportedAlgorithm() {
    for (const [algorithm, params] of Object.entries(WEB_CRYPTO_ALGORITHM_CONFIGS)) {
      try {
        await crypto.subtle.generateKey(params.generateParams, false, ["sign", "verify"]);
        return algorithm;
      } catch {
        continue;
      }
    }
    throw new Error(
      "Your browser does not support the encryption needed for secure authentication. Please use a different browser or upgrade to a supported browser."
    );
  }
  /**
   * Get the public key information
   */
  getKeyInfo() {
    return this.activeKeyPairRecord?.keyInfo || null;
  }
  /**
   * Reset the key pair by generating a new one
   */
  async resetKeyPair() {
    await this.clearStoredKeys();
    this.activeKeyPairRecord = await this.generateAndStoreNewKeyPair("active");
    this.pendingKeyPairRecord = null;
    return this.activeKeyPairRecord.keyInfo;
  }
  /**
   * Create X-Phantom-Stamp header value using stored private key
   * @param params - Parameters object with data and optional type/options
   * @returns Complete X-Phantom-Stamp header value
   */
  async stamp(params) {
    const { data } = params;
    if (!this.activeKeyPairRecord) {
      throw new Error("Stamper not initialized. Call init() first.");
    }
    const dataBytes = new Uint8Array(data);
    const signature = await crypto.subtle.sign(
      WEB_CRYPTO_ALGORITHM_CONFIGS[this.algorithm].signParams,
      this.activeKeyPairRecord.keyPair.privateKey,
      dataBytes
    );
    const signatureBase64url = (0, import_base64url.base64urlEncode)(new Uint8Array(signature));
    const stampType = params.type || this.type;
    const idToken = params.type === "OIDC" ? params.idToken : this.idToken;
    const salt = params.type === "OIDC" ? params.salt : this.salt;
    const stampData = stampType === "PKI" ? {
      // Decode base58 public key to bytes, then encode as base64url (consistent with ApiKeyStamper)
      publicKey: (0, import_base64url.base64urlEncode)(import_bs58.default.decode(this.activeKeyPairRecord.keyInfo.publicKey)),
      signature: signatureBase64url,
      kind: "PKI",
      algorithm: this.algorithm
    } : {
      kind: "OIDC",
      idToken,
      publicKey: (0, import_base64url.base64urlEncode)(import_bs58.default.decode(this.activeKeyPairRecord.keyInfo.publicKey)),
      salt,
      algorithm: this.algorithm,
      signature: signatureBase64url
    };
    const stampJson = JSON.stringify(stampData);
    return (0, import_base64url.base64urlEncode)(stampJson);
  }
  /**
   * Clear all stored keys
   */
  async clear() {
    await this.clearStoredKeys();
    this.activeKeyPairRecord = null;
    this.pendingKeyPairRecord = null;
  }
  async clearStoredKeys() {
    if (!this.db) {
      await this.openDB();
    }
    return new Promise((resolve, reject) => {
      const transaction = this.db.transaction([this.storeName], "readwrite");
      const store = transaction.objectStore(this.storeName);
      const deleteActiveKeyPair = store.delete(`${this.keyName}-active`);
      const deletePendingKeyPair = store.delete(`${this.keyName}-pending`);
      let completed = 0;
      const total = 2;
      const checkComplete = () => {
        completed++;
        if (completed === total) {
          resolve();
        }
      };
      deleteActiveKeyPair.onsuccess = checkComplete;
      deletePendingKeyPair.onsuccess = checkComplete;
      deleteActiveKeyPair.onerror = () => reject(deleteActiveKeyPair.error);
      deletePendingKeyPair.onerror = () => reject(deletePendingKeyPair.error);
    });
  }
  async openDB() {
    return new Promise((resolve, reject) => {
      const request = indexedDB.open(this.dbName, 1);
      request.onerror = () => reject(request.error);
      request.onsuccess = () => {
        this.db = request.result;
        resolve();
      };
      request.onupgradeneeded = (event) => {
        const db = event.target.result;
        if (!db.objectStoreNames.contains(this.storeName)) {
          db.createObjectStore(this.storeName);
        }
      };
    });
  }
  /**
   * Generate a new keypair for rotation without making it active
   */
  async rotateKeyPair() {
    if (!this.db) {
      await this.openDB();
    }
    this.pendingKeyPairRecord = await this.generateAndStoreNewKeyPair("pending");
    return this.pendingKeyPairRecord.keyInfo;
  }
  /**
   * Switch to the pending keypair, making it active and cleaning up the old one
   */
  async commitRotation(authenticatorId) {
    if (!this.pendingKeyPairRecord) {
      throw new Error("No pending keypair to commit");
    }
    if (this.activeKeyPairRecord) {
      await this.removeKeyPairRecord("active");
    }
    this.pendingKeyPairRecord.status = "active";
    this.pendingKeyPairRecord.authenticatorId = authenticatorId;
    this.pendingKeyPairRecord.keyInfo.authenticatorId = authenticatorId;
    this.activeKeyPairRecord = this.pendingKeyPairRecord;
    this.pendingKeyPairRecord = null;
    await this.storeKeyPairRecord(this.activeKeyPairRecord, "active");
    await this.removeKeyPairRecord("pending");
  }
  /**
   * Discard the pending keypair on rotation failure
   */
  async rollbackRotation() {
    if (!this.pendingKeyPairRecord) {
      return;
    }
    await this.removeKeyPairRecord("pending");
    this.pendingKeyPairRecord = null;
  }
  async generateAndStoreNewKeyPair(type) {
    const keyPair = await crypto.subtle.generateKey(
      WEB_CRYPTO_ALGORITHM_CONFIGS[this.algorithm].generateParams,
      false,
      // non-extractable - private key can never be exported
      ["sign", "verify"]
    );
    const rawPublicKeyBuffer = await crypto.subtle.exportKey("raw", keyPair.publicKey);
    const publicKeyBase58 = import_bs58.default.encode(new Uint8Array(rawPublicKeyBuffer));
    const keyIdBuffer = await crypto.subtle.digest("SHA-256", rawPublicKeyBuffer);
    const keyId = (0, import_base64url.base64urlEncode)(new Uint8Array(keyIdBuffer)).substring(0, 16);
    const now = Date.now();
    const keyInfo = {
      keyId,
      publicKey: publicKeyBase58,
      createdAt: now
    };
    const record = {
      keyPair,
      keyInfo,
      createdAt: now,
      expiresAt: 0,
      // Not used anymore, kept for backward compatibility
      status: type
    };
    await this.storeKeyPairRecord(record, type);
    return record;
  }
  async storeKeyPairRecord(record, type) {
    if (!this.db) {
      throw new Error("Database not initialized");
    }
    return new Promise((resolve, reject) => {
      const transaction = this.db.transaction([this.storeName], "readwrite");
      const store = transaction.objectStore(this.storeName);
      const request = store.put(record, `${this.keyName}-${type}`);
      request.onsuccess = () => resolve();
      request.onerror = () => reject(request.error);
    });
  }
  async loadActiveKeyPairRecord() {
    if (!this.db) {
      return null;
    }
    return new Promise((resolve, reject) => {
      const transaction = this.db.transaction([this.storeName], "readonly");
      const store = transaction.objectStore(this.storeName);
      const request = store.get(`${this.keyName}-active`);
      request.onsuccess = () => resolve(request.result || null);
      request.onerror = () => reject(request.error);
    });
  }
  async loadPendingKeyPairRecord() {
    if (!this.db) {
      return null;
    }
    return new Promise((resolve, reject) => {
      const transaction = this.db.transaction([this.storeName], "readonly");
      const store = transaction.objectStore(this.storeName);
      const request = store.get(`${this.keyName}-pending`);
      request.onsuccess = () => resolve(request.result || null);
      request.onerror = () => reject(request.error);
    });
  }
  async removeKeyPairRecord(type) {
    if (!this.db) {
      return;
    }
    return new Promise((resolve, reject) => {
      const transaction = this.db.transaction([this.storeName], "readwrite");
      const store = transaction.objectStore(this.storeName);
      const request = store.delete(`${this.keyName}-${type}`);
      request.onsuccess = () => resolve();
      request.onerror = () => reject(request.error);
    });
  }
};
