// @phantom/openapi-wallet-service@0.1.11/api/kmsrpcapi.ts
/* tslint:disable */
/* eslint-disable */
/**
 * kms-api
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import type { Configuration } from "../configuration";
import type { AxiosPromise, AxiosInstance, RawAxiosRequestConfig } from "axios";
import globalAxios from "axios";
// Some imports not used depending on template conditions
// @ts-ignore
import {
  DUMMY_BASE_URL,
  assertParamExists,
  setApiKeyToObject,
  setBasicAuthToObject,
  setBearerAuthToObject,
  setOAuthToObject,
  setSearchParams,
  serializeDataIfNeeded,
  toPathString,
  createRequestFunction,
} from "../common";
// @ts-ignore
import { BASE_PATH, COLLECTION_FORMATS, type RequestArgs, BaseAPI, RequiredError, operationServerMap } from "../base";
// @ts-ignore
import type { ErrorResponse } from "../model";
// @ts-ignore
import type { KmsRpcRequest } from "../model";
// @ts-ignore
import type { KmsRpcResponseV2 } from "../model";
/**
 * KMSRPCApi - axios parameter creator
 * @export
 */
export const KMSRPCApiAxiosParamCreator = function (configuration?: Configuration) {
  return {
    /**
     *  ### Endpoint This endpoint implements an RPC interface for KMS operations. The request body should be a JSON object with the following fields: - `method`: The RPC method to call. Supported methods:   - `createOrganization`: Creates a new organization with users and authenticators   - `createAuthenticator`: Creates a new authenticator for a user in an organization   - `deleteAuthenticator`: Deletes an authenticator from a user in an organization   - `createWallet`: Creates a new wallet within an organization   - `getOrganizationWallets`: Retrieves all wallets owned by an organization with pagination support   - `grantOrganizationAccess`: Grants wallet access to another organization   - `getOrCreatePhantomOrganization`: Gets or creates a phantom organization with the given public key and algorithm   - `signRawPayload`: Signs raw bytes with a specified wallet and derivation path   - `signTransaction`: Signs a transaction with a specified wallet and derivation path   - `initMnemonicImport`: Initializes a mnemonic import for a user in an organization   - `completeMnemonicImport`: Completes the mnemonic import process for a user in an organization   - `exportWallet`: Exports a wallet\'s data in an encrypted format - `params`: Parameters for the method, specific to each method type - `timestampMs`: Request timestamp in milliseconds (must be within 5 minutes of current time) ### Response The response body will be a JSON object with the following fields: - `result`: The result of the RPC call. The structure depends on the method:   - For `createOrganization`: Contains organization details   - For `createAuthenticator`: Contains authenticator details   - For `createWallet`: Contains wallet ID and name   - For `getOrganizationWallets`: Contains array of wallets with wallet_id and wallet_name, plus total_count, limit, and offset for pagination   - For `grantOrganizationAccess`: Contains confirmation of access grant with target organization ID and wallet ID   - For `getOrCreatePhantomOrganization`: Contains organization details for the phantom organization   - For `signRawPayload`: Contains signature and public key   - For `signTransaction`: Contains signed transaction and public key   - For `initMnemonicImport`: Contains import envelope and public key   - For `completeMnemonicImport`: Contains wallet ID and name   - For `exportWallet`: Contains encrypted wallet data
     * @param {KmsRpcRequest} kmsRpcRequest
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     */
    postKmsRpc: async (kmsRpcRequest: KmsRpcRequest, options: RawAxiosRequestConfig = {}): Promise<RequestArgs> => {
      // verify required parameter 'kmsRpcRequest' is not null or undefined
      assertParamExists("postKmsRpc", "kmsRpcRequest", kmsRpcRequest);
      const localVarPath = `/kms/rpc`;
      // use dummy base URL string because the URL constructor only accepts absolute URLs.
      const localVarUrlObj = new URL(localVarPath, DUMMY_BASE_URL);
      let baseOptions;
      if (configuration) {
        baseOptions = configuration.baseOptions;
      }

      const localVarRequestOptions = { method: "POST", ...baseOptions, ...options };
      const localVarHeaderParameter = {} as any;
      const localVarQueryParameter = {} as any;

      localVarHeaderParameter["Content-Type"] = "application/json";

      setSearchParams(localVarUrlObj, localVarQueryParameter);
      let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
      localVarRequestOptions.headers = { ...localVarHeaderParameter, ...headersFromBaseOptions, ...options.headers };
      localVarRequestOptions.data = serializeDataIfNeeded(kmsRpcRequest, localVarRequestOptions, configuration);

      return {
        url: toPathString(localVarUrlObj),
        options: localVarRequestOptions,
      };
    },
  };
};

/**
 * KMSRPCApi - functional programming interface
 * @export
 */
export const KMSRPCApiFp = function (configuration?: Configuration) {
  const localVarAxiosParamCreator = KMSRPCApiAxiosParamCreator(configuration);
  return {
    /**
     *  ### Endpoint This endpoint implements an RPC interface for KMS operations. The request body should be a JSON object with the following fields: - `method`: The RPC method to call. Supported methods:   - `createOrganization`: Creates a new organization with users and authenticators   - `createAuthenticator`: Creates a new authenticator for a user in an organization   - `deleteAuthenticator`: Deletes an authenticator from a user in an organization   - `createWallet`: Creates a new wallet within an organization   - `getOrganizationWallets`: Retrieves all wallets owned by an organization with pagination support   - `grantOrganizationAccess`: Grants wallet access to another organization   - `getOrCreatePhantomOrganization`: Gets or creates a phantom organization with the given public key and algorithm   - `signRawPayload`: Signs raw bytes with a specified wallet and derivation path   - `signTransaction`: Signs a transaction with a specified wallet and derivation path   - `initMnemonicImport`: Initializes a mnemonic import for a user in an organization   - `completeMnemonicImport`: Completes the mnemonic import process for a user in an organization   - `exportWallet`: Exports a wallet\'s data in an encrypted format - `params`: Parameters for the method, specific to each method type - `timestampMs`: Request timestamp in milliseconds (must be within 5 minutes of current time) ### Response The response body will be a JSON object with the following fields: - `result`: The result of the RPC call. The structure depends on the method:   - For `createOrganization`: Contains organization details   - For `createAuthenticator`: Contains authenticator details   - For `createWallet`: Contains wallet ID and name   - For `getOrganizationWallets`: Contains array of wallets with wallet_id and wallet_name, plus total_count, limit, and offset for pagination   - For `grantOrganizationAccess`: Contains confirmation of access grant with target organization ID and wallet ID   - For `getOrCreatePhantomOrganization`: Contains organization details for the phantom organization   - For `signRawPayload`: Contains signature and public key   - For `signTransaction`: Contains signed transaction and public key   - For `initMnemonicImport`: Contains import envelope and public key   - For `completeMnemonicImport`: Contains wallet ID and name   - For `exportWallet`: Contains encrypted wallet data
     * @param {KmsRpcRequest} kmsRpcRequest
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     */
    async postKmsRpc(
      kmsRpcRequest: KmsRpcRequest,
      options?: RawAxiosRequestConfig,
    ): Promise<(axios?: AxiosInstance, basePath?: string) => AxiosPromise<KmsRpcResponseV2>> {
      const localVarAxiosArgs = await localVarAxiosParamCreator.postKmsRpc(kmsRpcRequest, options);
      const localVarOperationServerIndex = configuration?.serverIndex ?? 0;
      const localVarOperationServerBasePath =
        operationServerMap["KMSRPCApi.postKmsRpc"]?.[localVarOperationServerIndex]?.url;
      return (axios, basePath) =>
        createRequestFunction(
          localVarAxiosArgs,
          globalAxios,
          BASE_PATH,
          configuration,
        )(axios, localVarOperationServerBasePath || basePath);
    },
  };
};

/**
 * KMSRPCApi - factory interface
 * @export
 */
export const KMSRPCApiFactory = function (configuration?: Configuration, basePath?: string, axios?: AxiosInstance) {
  const localVarFp = KMSRPCApiFp(configuration);
  return {
    /**
     *  ### Endpoint This endpoint implements an RPC interface for KMS operations. The request body should be a JSON object with the following fields: - `method`: The RPC method to call. Supported methods:   - `createOrganization`: Creates a new organization with users and authenticators   - `createAuthenticator`: Creates a new authenticator for a user in an organization   - `deleteAuthenticator`: Deletes an authenticator from a user in an organization   - `createWallet`: Creates a new wallet within an organization   - `getOrganizationWallets`: Retrieves all wallets owned by an organization with pagination support   - `grantOrganizationAccess`: Grants wallet access to another organization   - `getOrCreatePhantomOrganization`: Gets or creates a phantom organization with the given public key and algorithm   - `signRawPayload`: Signs raw bytes with a specified wallet and derivation path   - `signTransaction`: Signs a transaction with a specified wallet and derivation path   - `initMnemonicImport`: Initializes a mnemonic import for a user in an organization   - `completeMnemonicImport`: Completes the mnemonic import process for a user in an organization   - `exportWallet`: Exports a wallet\'s data in an encrypted format - `params`: Parameters for the method, specific to each method type - `timestampMs`: Request timestamp in milliseconds (must be within 5 minutes of current time) ### Response The response body will be a JSON object with the following fields: - `result`: The result of the RPC call. The structure depends on the method:   - For `createOrganization`: Contains organization details   - For `createAuthenticator`: Contains authenticator details   - For `createWallet`: Contains wallet ID and name   - For `getOrganizationWallets`: Contains array of wallets with wallet_id and wallet_name, plus total_count, limit, and offset for pagination   - For `grantOrganizationAccess`: Contains confirmation of access grant with target organization ID and wallet ID   - For `getOrCreatePhantomOrganization`: Contains organization details for the phantom organization   - For `signRawPayload`: Contains signature and public key   - For `signTransaction`: Contains signed transaction and public key   - For `initMnemonicImport`: Contains import envelope and public key   - For `completeMnemonicImport`: Contains wallet ID and name   - For `exportWallet`: Contains encrypted wallet data
     * @param {KmsRpcRequest} kmsRpcRequest
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     */
    postKmsRpc(kmsRpcRequest: KmsRpcRequest, options?: RawAxiosRequestConfig): AxiosPromise<KmsRpcResponseV2> {
      return localVarFp.postKmsRpc(kmsRpcRequest, options).then(request => request(axios, basePath));
    },
  };
};

/**
 * KMSRPCApi - object-oriented interface
 * @export
 * @class KMSRPCApi
 * @extends {BaseAPI}
 */
export class KMSRPCApi extends BaseAPI {
  /**
   *  ### Endpoint This endpoint implements an RPC interface for KMS operations. The request body should be a JSON object with the following fields: - `method`: The RPC method to call. Supported methods:   - `createOrganization`: Creates a new organization with users and authenticators   - `createAuthenticator`: Creates a new authenticator for a user in an organization   - `deleteAuthenticator`: Deletes an authenticator from a user in an organization   - `createWallet`: Creates a new wallet within an organization   - `getOrganizationWallets`: Retrieves all wallets owned by an organization with pagination support   - `grantOrganizationAccess`: Grants wallet access to another organization   - `getOrCreatePhantomOrganization`: Gets or creates a phantom organization with the given public key and algorithm   - `signRawPayload`: Signs raw bytes with a specified wallet and derivation path   - `signTransaction`: Signs a transaction with a specified wallet and derivation path   - `initMnemonicImport`: Initializes a mnemonic import for a user in an organization   - `completeMnemonicImport`: Completes the mnemonic import process for a user in an organization   - `exportWallet`: Exports a wallet\'s data in an encrypted format - `params`: Parameters for the method, specific to each method type - `timestampMs`: Request timestamp in milliseconds (must be within 5 minutes of current time) ### Response The response body will be a JSON object with the following fields: - `result`: The result of the RPC call. The structure depends on the method:   - For `createOrganization`: Contains organization details   - For `createAuthenticator`: Contains authenticator details   - For `createWallet`: Contains wallet ID and name   - For `getOrganizationWallets`: Contains array of wallets with wallet_id and wallet_name, plus total_count, limit, and offset for pagination   - For `grantOrganizationAccess`: Contains confirmation of access grant with target organization ID and wallet ID   - For `getOrCreatePhantomOrganization`: Contains organization details for the phantom organization   - For `signRawPayload`: Contains signature and public key   - For `signTransaction`: Contains signed transaction and public key   - For `initMnemonicImport`: Contains import envelope and public key   - For `completeMnemonicImport`: Contains wallet ID and name   - For `exportWallet`: Contains encrypted wallet data
   * @param {KmsRpcRequest} kmsRpcRequest
   * @param {*} [options] Override http request option.
   * @throws {RequiredError}
   * @memberof KMSRPCApi
   */
  public postKmsRpc(kmsRpcRequest: KmsRpcRequest, options?: RawAxiosRequestConfig) {
    return KMSRPCApiFp(this.configuration)
      .postKmsRpc(kmsRpcRequest, options)
      .then(request => request(this.axios, this.basePath));
  }
}
